// bobwai--app-header-button/src/data.ts
import * as b from "bobril";
import * as Color from "bobwai--color";
import * as FormElements from "bobwai--form-elements";
import * as Focus from "bobwai--focus";
import * as TooltipEmplacer from "bobwai--tooltip-emplacer";

export interface IData {
    type?: Type;
    tooltip?: FormElements.ITooltipData;
    tooltipPreferredPlacement?: TooltipEmplacer.Placement;
    preIconTitle?: string;
    title?: string;
    isSelected?: boolean;
    icon?: b.IBobrilNode;
    hoveredIcon?: b.IBobrilNode;
    rightIcon?: b.IBobrilNode;
    rightHoveredIcon?: b.IBobrilNode;
    isDisabled?: boolean;
    customTheme?: ITheme;
    padding?: Padding;
    siblingButtons?: SiblingButtons;
    theme?: Theme;
    url?: string;
    webContentAccessibility?: WebContentAccessibility;
    onClick?: () => void;
    onMouseEnter?: () => void;
    onMouseLeave?: () => void;
}

export type WebContentAccessibility = Required<Pick<Focus.IWebContentAccessibility, "ariaLabel">>;

export interface ITheme {
    color?: Color.IColor;
    hoverColor?: Color.IColor;
    backgroundColor?: Color.IColor;
    selectedBackgroundColor?: Color.IColor;
}

export enum Padding {
    Default,
    Small,
}

export enum SiblingButtons {
    None,
    Left,
    Right,
    Both,
}

export enum Type {
    Navigation,
    Action,
}

export enum Theme {
    Default,
    Light,
}
