// bobwai--app-header-button/src/publicConstants.ts
import * as Colors from "bobwai--color";

export const ButtonBackgroundActive: Colors.IColor = Colors.Basic; //OK
export const ButtonBackgroundOver: Colors.IColor = Colors.White_08; //OK
export const ButtonBackgroundSelected: Colors.IColor = Colors.Application; //OK
// does not make sence - when it is selected it should have no hover style
//export const ButtonBackgroundSelectedOver: Colors.IColor = Colors.Application;
export const TitleColorActive = Colors.White_75; //OK
export const TitleColorSelected = Colors.White; //OK

export const ButtonBackgroundActiveLight: Colors.IColor = Colors.White; //OK
export const ButtonBackgroundOverLight: Colors.IColor = Colors.Basic_08; //OK

// same as ButtonBackgroundOverLight
// export const ButtonHoverBackgroundLight: Colors.IColor = Colors.Grey300;
export const TitleColorActiveLight: Colors.IColor = Colors.Basic_80; //OK
export const TitleColorSelectedLight: Colors.IColor = Colors.Basic; //OK

let propEnableRTL = false;
export function EnableRTL(): boolean {
    return propEnableRTL;
}

export function setEnableRTL(value: boolean): void {
    propEnableRTL = value;
}
