// bobwai--textarea-tagged/src/diff.ts
export interface Diff {
    start: number;
    removed: number;
    added: number;
    insertion?: string;
}

// modified from https://stackoverflow.com/a/29402724
export function diffStrings(oldText: string, newText: string): Diff {
    // Find the index at which the change began
    var start = 0;
    while (start < oldText.length && start < newText.length && oldText[start] == newText[start]) {
        start++;
    }
    // Find the index at which the change ended (relative to the end of the string)
    var end = 0;
    while (
        end < oldText.length &&
        end < newText.length &&
        oldText.length - end > start &&
        newText.length - end > start &&
        oldText[oldText.length - 1 - end] == newText[newText.length - 1 - end]
    ) {
        end++;
    }
    // The change end of the new string (ne) and old string (oe)
    var ne = newText.length - end;
    var oe = oldText.length - end;
    // The number of chars removed and added
    var removed = oe - start;
    var added = ne - start;
    var insertion;
    if (added) {
        insertion = newText.substr(start, added);
    }
    return { start, removed, added, insertion };
}
