// bobwai--textarea-tagged/src/style.ts
import * as b from "bobril";
import { ICustomTheme } from "bobwai--textarea";
import * as font from "bobwai--font";
import * as color from "bobwai--color";
import * as mixins from "bobwai--style-mixins";

export const avatarSize = 32;
export const suggestionBoxWidth = 200;
export const suggestionItemHeight = 50;
const suggestionItemPadding = 10;
const avatarMargin = 15;
const textWidth = suggestionBoxWidth - 2 * suggestionItemPadding - avatarSize - avatarMargin;
export const offsetFromViewportEdge = 20;

export const textareaDefaultStyle: ICustomTheme = {
    fontFamily: font.FontFamily,
    fontSize: parseInt(font.FontSizeDefault),
    padding: "7px 12px",
    lineHeight: "19px"
};

export const containerStyle = b.styleDef({
    overflow: "hidden",
    position: "relative"
});

export const shadowStyle = b.styleDef({
    pointerEvents: "none",
    position: "absolute",
    top: 1,
    right: 1,
    left: 1,
    whiteSpace: "pre-wrap",
    wordBreak: "break-word"
});

export const wrapperStyle = b.styleDef({
    position: "relative"
});

export const listContainerStyle = b.styleDef({
    backgroundColor: color.White,
    width: suggestionBoxWidth,
    userSelect: "none",
    boxShadow: "0 4px 8px rgba(0,0,0,0.3)"
});

export const listItemHoverStyle: b.IBobrilStyle = {
    backgroundColor: color.Grey200
};

export const listItemStyle = b.styleDef(
    {
        boxSizing: "border-box",
        height: suggestionItemHeight,
        padding: suggestionItemPadding,
        display: "flex",
        alignItems: "center",
        borderBottom: `1px solid ${color.Grey300}`,
        cursor: "pointer"
    },
    {
        hover: listItemHoverStyle,
        "last-child": {
            borderBottom: "none"
        }
    }
);

export const itemBodyStyle = b.styleDef({
    display: "flex",
    flexDirection: "column",
    marginLeft: avatarMargin
});

export const nameStyle = b.styleDef([
    mixins.lineEllipsis(),
    {
        fontFamily: font.FontFamily,
        fontWeight: font.FontWeightSemibold,
        fontSize: 13,
        color: color.Basic,
        width: textWidth
    }
]);

export const emailStyle = b.styleDef([
    mixins.lineEllipsis(),
    {
        fontFamily: font.FontFamily,
        fontSize: 10,
        color: color.Basic,
        width: textWidth
    }
]);

export const moreItemStyle = b.styleDef({
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    textAlign: "center",
    fontFamily: font.FontFamily,
    fontStyle: "italic",
    fontSize: 13,
    color: color.Grey500,
    padding: suggestionItemPadding,
    cursor: "default"
});

export const hiddenStyle = b.styleDef({
    opacity: 0,
    visibility: "hidden"
});
