// helpers--web-content-accessibility/src/webContentAccessibilityAttributes.ts
export namespace AriaAttributeName {
    export const label: string = "aria-label";
    export const disabled: string = "aria-disabled";
    export const pressed: string = "aria-pressed";
    export const hasPopup: string = "aria-haspopup";
    export const expanded: string = "aria-expanded";
    export const sort: string = "aria-sort";
    export const role: string = "role";
    export const isModal: string = "aria-modal";
    export const live: string = "aria-live";
    export const selected: string = "aria-selected";
    export const checked: string = "aria-checked";
    export const multiline: string = "aria-multiline";
    export const hidden: string = "aria-hidden";
    export const activeListItemId: string = "aria-activedescendant";
    export const controls: string = "aria-controls";
    export const autocomplete: string = "aria-autocomplete";
    export const id: string = "id";
    export const level: string = "aria-level";
    export const setSize: string = "aria-setsize";
    export const positionSet: string = "aria-posinset";
    export const current: string = "aria-current";
    export const invalid: string = "aria-invalid";
    export const valueMax: string = "aria-valuemax";
    export const valueMin: string = "aria-valuemin";
    export const valueNow: string = "aria-valuenow";
    export const required: string = "aria-required";
    export const alternativeDescription: string = "alt";
    export const describedBy: string = "aria-describedby";
    export const labelledBy: string = "aria-labelledby";
    export const labelFor: string = "for";
}

export namespace AriaAttributeValue {
    export const hasPopupListbox: string = "listbox";
    export const hasPopupTree: string = "tree";
    export const autocompleteList: string = "list";
    export const currentDate: string = "date";
}
