// bobwai--animation-frame/examples/src/mockComponent.ts
import * as b from "bobril";
import * as Button from "bobwai--button";
import * as Animation from "bobwai--animation";
import * as animationFrame from "../../src/lib";

export interface IData {
    content: b.IBobrilNode;
}

interface IMockContext extends b.IBobrilCtx {
    data: IData;
    isElementHiding: boolean | undefined;
    isElementVisible: boolean | undefined;
}

export let create = b.createComponent<IData>({
    id: "bobwai--mock-component",
    render(ctx: IMockContext, me: b.IBobrilNode): void {
        me.children = [
            b.style(
                {
                    tag: "div",
                    children: Button.create({
                        label: "Hover to show element",
                        onClick: () => {}
                    }),
                    component: {
                        onMouseEnter(_c: b.IBobrilCtx, _e: b.IBobrilMouseEvent): boolean {
                            if (ctx.isElementHiding) {
                                ctx.isElementHiding = false;
                            }
                            ctx.isElementVisible = true;
                            b.invalidate(ctx);
                            return false;
                        },
                        onMouseLeave(_c: b.IBobrilCtx, _e: b.IBobrilMouseEvent): boolean {
                            ctx.isElementHiding = true;
                            b.invalidate(ctx);
                            return false;
                        }
                    }
                },
                { display: "inline-block" }
            ),
            ctx.isElementVisible &&
                animationFrame.create({
                    content: ctx.data.content,
                    isHiding: ctx.isElementHiding,
                    showHideAnimation: Animation.fadeInFadeOutAnimation(),
                    onHideAnimationFinished: () => {
                        ctx.isElementVisible = undefined;
                        ctx.isElementHiding = undefined;
                        b.invalidate(ctx);
                    }
                })
        ];
    }
});
