// bobwai--chat/examples/src/helpers.ts
import * as b from "bobril";
import * as Icon from "bobwai--icon";
import * as color from "bobwai--color";
import * as button from "bobwai--button";

export const iconRdComment = b.styledDiv(null, b.sprite(Icon.user_medium_png, color.Warning));
export const iconRdReply = b.styledDiv(null, b.sprite(Icon.user_small_24_png, color.Warning));
export const iconJdComment = b.styledDiv(null, b.sprite(Icon.user_medium_png, color.Error));
export const iconJdReply = b.styledDiv(null, b.sprite(Icon.user_small_24_png, color.Error));
export const iconMdComment = b.styledDiv(null, b.sprite(Icon.user_medium_png, color.Application));
export const iconMdReply = b.styledDiv(null, b.sprite(Icon.user_small_24_png, color.Application));
export const iconAnonymousComment = b.styledDiv(null, b.sprite(Icon.user_medium_png, color.White));
export const iconAnonymousReply = b.styledDiv(null, b.sprite(Icon.user_small_24_png, color.White));

export function createViewAreaButton(label: string, onClick: () => void) {
    return button.create({
        leftIcon: b.styledDiv(null, b.sprite(Icon.visible_small_png, color.Basic), { display: "inline-block" }),
        leftIconHover: b.styledDiv(null, b.sprite(Icon.visible_small_png, color.Application), { display: "inline-block" }),
        onClick: onClick,
        label: label,
        size: button.Size.Small,
        color: button.Theme.None
    });
}

export function createAddAreaButton(onClick: () => void, label: string = "ADD AREA") {
    return button.create({
        leftIcon: b.styledDiv(null, b.sprite(Icon.resizable_area_small_png, color.Basic), { display: "inline-block" }),
        leftIconHover: b.styledDiv(null, b.sprite(Icon.resizable_area_small_png, color.Application), { display: "inline-block" }),
        onClick: onClick,
        label: label,
        size: button.Size.Small,
        color: button.Theme.None
    });
}
