// helpers--rtl-text/src/lib.ts
import * as b from "bobril";
import * as FormElements from "bobwai--form-elements";
import * as s from "./styles";

export interface IRtlTextParams {
    enableRTL?: () => boolean;
    text?: string;
    textAlign?: FormElements.TextAlign;
}

export function getRtlTextStyles(params: IRtlTextParams): b.IBobrilStyles {
    let isRtlEnabled = (!params.enableRTL || params.enableRTL()) && isRTL(params.text);
    if (!isRtlEnabled) return [];

    return [s.textDirectionRtlStyle, params.textAlign !== undefined && getTextAlignStyle(params.textAlign)];
}

function getTextAlignStyle(textAlign: FormElements.TextAlign): b.IBobrilStyleDef {
    switch (textAlign) {
        case FormElements.TextAlign.Left:
            return s.rtlTextAlignLeftStyle;
        case FormElements.TextAlign.Center:
            return s.rtlTextAlignCenterStyle;
        case FormElements.TextAlign.Right:
            return s.rtlTextAlignRightStyle;
    }
}

function isRTL(text?: string): boolean {
    if (!text) return false;

    let rtlCharacters = /[\u0590-\u05FF\u0600-\u0659\u0670-\u06FF\u0750-\u077F\u08A0-\u08FF\uFB50-\uFDFF\uFE70-\uFEFF]/;
    return rtlCharacters.test(text);
}
