// bobwai--style-mixins/src/shadowStyles.ts
import * as b from "bobril";
import * as Color from "bobwai--color";

export namespace shadowStylesMixins {
    /** Deprecated: It will be removed in future version - use getDropShadow100 instead of this */
    export function getShadow1Styles(): b.CSSInlineStyles {
        return getDropShadow100();
    }
    /** Deprecated: It will be removed in future version - use getDropShadow200 instead of this  */
    export function getShadow3Styles(): b.CSSInlineStyles {
        return getDropShadow200();
    }
    /** Deprecated: It will be removed in future version - use getDropShadow300 instead of this  */
    export function getShadow4Styles(): b.CSSInlineStyles {
        return getDropShadow300();
    }
    /** Deprecated: It will be removed in future version - use getDropShadow400 instead of this */
    export function getShadow6Styles(): b.CSSInlineStyles {
        return getDropShadow400();
    }

    export function getDropShadow100(): b.CSSInlineStyles {
        return {
            boxShadow: `0px 2px 3px 0 ${Color.Basic_Opacity300()}, 0px 1px 4px 0 ${Color.Basic_Opacity200()}`,
        };
    }
    export function getDropShadow200(): b.CSSInlineStyles {
        return {
            boxShadow: `0px 5px 7px 0 ${Color.Basic_Opacity300()}, 0px 2px 10px 0 ${Color.Basic_Opacity200()}`,
        };
    }
    export function getDropShadow300(): b.CSSInlineStyles {
        return {
            boxShadow: `0px 7px 10px 0 ${Color.Basic_Opacity300()}, 0px 3px 13px 0 ${Color.Basic_Opacity200()}`,
        };
    }
    export function getDropShadow400(): b.CSSInlineStyles {
        return {
            boxShadow: `0px 10px 15px 0 ${Color.Basic_Opacity300()}, 0px 4px 20px 0 ${Color.Basic_Opacity200()}`,
        };
    }
}
