// bobwai--viewer-container/src/mockComponents/header.ts
import * as b from "bobril";
import * as MockColor from "./color";

interface IData {
    content?: string;
    actionContent?: b.IBobrilChild;
}

interface IContext extends b.IBobrilCtx {
    data: IData;
}

export let create = b.createComponent<IData>({
    render(ctx: IContext, me: b.IBobrilNode) {
        me.children = [
            b.styledDiv(ctx.data.content, viewerContainerHeaderTitleStyle),
            ctx.data.actionContent && b.styledDiv(ctx.data.actionContent, viewerContainerActionStyle)
        ];
        b.style(me, viewerContainerHeaderStyle);
    }
});

const viewerContainerHeaderTitleStyle = b.styleDef({
    minWidth: 1,
    fontSize: 22,
    color: MockColor.baseColor,
    width: "100%",
    overflow: "hidden",
    textOverflow: "ellipsis",
    whiteSpace: "nowrap"
});

const viewerContainerHeaderStyle = b.styleDef({
    lineHeight: "30px",
    display: "flex"
});

const viewerContainerActionStyle = b.styleDef({
    marginLeft: 5,
    display: "flex",
    alignItems: "center",
    flexShrink: 0,
    whiteSpace: "nowrap"
});
