// bobwai--l-view-app-breadcrumb/src/lib.ts
import * as b from "bobril";
import * as styles from "./styles";
import * as HorizontalScrollCorrection from "bobwai--horizontal-scroll-correction";

export let create = b.createVirtualComponent<IData>({
    id: "bobwai--l-view-app-breadcrumb",
    render(ctx: IContext, me: b.IBobrilNode): void {
        me.children = HorizontalScrollCorrection.create({
            content: b.styledDiv(
                ctx.data.children,
                styles.contentStyle,
                ctx.data.paddingSize === PaddingSize.Small && styles.contentSmallPaddingStyle,
                ctx.data.isCombinedWithAppHeader && styles.combinedWithAppHeaderStyle
            )
        });
    }
});

export interface IData {
    children?: b.IBobrilChildren;
    isCombinedWithAppHeader: boolean;
    paddingSize?: PaddingSize;
}

export enum PaddingSize {
    Default,
    Small
}

export * from "./publicConstants";

interface IContext extends b.IBobrilCtx {
    data: IData;
}

// added by script - BEGIN
export const LViewAppBreadcrumb = create;
// added by script - END
