// bobwai--l-view-app-breadcrumb/src/styles.ts
import * as b from "bobril";
import * as constants from "./publicConstants";
import * as LViewAppHeader from "bobwai--l-view-app-header";
import * as L_App from "bobwai--l-app";

namespace localVariables {
    export const contentTop = LViewAppHeader.Height;
}

export const contentStyle = b.styleDef(
    {
        width: "100%",
        boxSizing: "border-box",
        height: `${constants.Height}px`,
        background: constants.BackgroundColor,
        paddingLeft: constants.ContentPadding,
        paddingRight: constants.ContentPadding,
        display: "inline-block",
        lineHeight: `${constants.Height}px`,
        verticalAlign: "middle",
        position: "fixed",
        minWidth: L_App.MinWidth,
        top: 0,
        borderBottomStyle: "solid",
        borderBottomWidth: 1,
        borderBottomColor: constants.BorderColor
    },
    {},
    "l-view-app-breadcrumb"
);

export const contentSmallPaddingStyle = b.styleDefEx(
    contentStyle,
    {
        paddingLeft: constants.ContentPaddingSmall,
        paddingRight: constants.ContentPaddingSmall
    },
    {},
    "l-view-app-breadcrumb-small-padding"
);

export const combinedWithAppHeaderStyle = b.styleDef({
    top: localVariables.contentTop
});
