// bobwai--chat/src/controlPanelStyle.ts
import * as b from "bobril";

const controlPanelStyle = b.styleDef(
    {
        height: 35
    },
    {},
    "bobwai--chat-control-panel"
);

export const leftControlPanelStyle = b.styleDef(
    [controlPanelStyle, { margin: 0, cssFloat: "left" }],
    {},
    "bobwai--chat-left-control-panel"
);

export const leftRootControlPanelStyle = b.styleDef(
    [leftControlPanelStyle, { marginLeft: 64 }],
    {},
    "bobwai--chat-left-control-panel--root"
);

export const leftReplyControlPanelStyle = b.styleDef(
    [leftControlPanelStyle, { marginLeft: 32 }],
    {},
    "bobwai--chat-left-control-panel--reply"
);

export const fullSizeControlPanelStyle = b.styleDef(
    [leftControlPanelStyle, { width: "100%" }],
    {},
    "bobwai--chat-full-control-panel--reply"
);

export const rightControlPanelStyle = b.styleDef(
    [controlPanelStyle, { cssFloat: "right", margin: 0 }],
    {},
    "bobwai--chat-right-control-panel"
);

export const rightRootControlPanelStyle = b.styleDef(
    [rightControlPanelStyle, { marginRight: 25 }],
    {},
    "bobwai--chat-right-control-panel--root"
);
