// bobwai--chat/src/editStyle.ts
import * as b from "bobril";
import * as color from "bobwai--color";

function textStyle(textAlign: string, cssFloat: string, marginLeft: number, marginRight: number) {
    return {
        textAlign: textAlign,
        color: color.Application,
        cssFloat: cssFloat,
        marginLeft: marginLeft,
        marginRight: marginRight
    };
}

export const editSubmitStyle = b.styleDef({ height: 40 }, {}, "bobwai--edit-submit-style");

export const editTextStyle = b.styleDef(textStyle("right", "right", 0, 0), {}, "bobwai--chat-edit-label");

export const deleteTextStyleFull = b.styleDef(textStyle("left", "left", 0, 0), {}, "bobwai--chat-delete-label");

export const deleteReplyTextStyleFull = b.styleDef(textStyle("left", "left", 50, 0), {}, "bobwai--chat-delete-with-reply-label");

export const deleteButtonReplyTextStyle = b.styleDef(textStyle("left", "left", 70, 0), {}, "bobwai--chat-delete-with-reply-label");

export const submitTextStyle = b.styleDef([textStyle("left", "right", 0, 0), { display: "inline-block" }], {}, "bobwai--chat-submit-label");

export const cancelDeleteTextStyle = b.styleDef(
    { textAlign: "left", color: color.Application, display: "inline-block" },
    {},
    "bobwai--chat-cancel-delete-label"
);
