// bobwai--chat/src/testSelectors.ts
export namespace Selectors {
    export const id = ".bobwai--chat";
    export const rootComment = "/~.bobwai--chat-root-comment";
    export const textArea = "/~.bobwai--textarea/textarea";
    export const rootSubmitButton = "/#button";
    export const comments = "/~.bobwai--chat-submited-comment";
    export const replyButton = "/~.bobwai--chat-reply-button/~.bobwai--button[0]";
    export const replyTextArea = "/~.bobwai--chat-reply/~textarea";
    export const replySubmitButton = "/~.bobwai--chat-reply/~.bobwai--button";
    export const replyEditButtons = "/~.bobwai--chat-edit-buttons";
    export const editOKbutton = "/~.bobwai--chat-submited-comment--text[last()]/~.bobwai--button[0]";
    export const editCancelbutton = "/~.bobwai--chat-submited-comment--text[last()]/~.bobwai--button[1]";
    export const editDeleteButton = "/~.bobwai--chat-edit-buttons/~.bobwai--button[0]";
    export const editEditButton = "/~.bobwai--chat-edit-buttons/~.bobwai--button[1]";
    export const lastCommentText = "/~.bobwai--chat-submited-comment--text[last()]/div[0]";
    export const getReplies = (commentId: number): string => {
        return `/~#${commentId}_rtb_replies_comments/~.bobwai--chat-submited-comment`;
    };
    export const getComment = (commentId: number): string => {
        return `/~#${commentId}`;
    };
}
