// bobwai--tab-item/src/publicConstants.ts
import * as Color from "bobwai--color";
import * as ColorOperations from "bobwai--color-operations";

export const BackgroundColor = Color.Grey200;
export const BackgroundColorActive = Color.White;
export const Height = 40;
export const SmallSizeHeight = 30;
export const TextPaddingRightLeft = 20;
export const MaxWidth = 250;
export const MarginLeftIcon = 15;
export const DeleteIconWidth = 20;

let propTextColor: Color.IColor = Color.Basic;
export function TextColor(): string {
    return ColorOperations.makeFixedValue(propTextColor);
}
export function setTextColor(val: Color.IColor): void {
    propTextColor = val;
}

let propTextColorActive: Color.IColor = Color.Application;
export function TextColorActive(): string {
    return ColorOperations.makeFixedValue(propTextColorActive);
}
export function setTextColorActive(val: Color.IColor): void {
    propTextColorActive = val;
}

let propUseLatestStyles: boolean = false;
export function UseLatestStyles(): boolean {
    return propUseLatestStyles;
}
export function setUseLatestStyles(val: boolean): void {
    propUseLatestStyles = val;
}
