// bobwai--app-header/src/styles.ts
import * as b from "bobril";
import * as constants from "./publicConstants";

export const baseStyle = b.styleDef(
    {
        display: "flex",
        background: constants.BackgroundColor,
        height: constants.Height,
        outline: "none",
        position: "relative",
        justifyContent: "space-between",
    },
    {},
    "app-header"
);

export const baseWithRightContentOnlyStyle = b.styleDefEx(
    baseStyle,
    {
        justifyContent: "flex-end",
    },
    {},
    "app-header-right-content-only"
);

export const bottomBorderStyle = b.styleDef({
    position: "absolute",
    left: 0,
    right: 0,
    height: 1,
    backgroundColor: constants.BorderColor,
    bottom: 0,
});

export const lightStyle = b.styleDef({
    background: constants.LightBackgroundColor,
});

export const dxbLightStyle = b.styleDef({
    background: constants.BackgroundColorDXBLight,
});

export const oms500Style = b.styleDef({
    background: "#2c2d2d",
});

export const leftStyle = b.styleDef({
    height: constants.Height,
    flexShrink: 0,
});
export const rightStyle = b.styleDef({
    flexShrink: 0,
    height: constants.Height,
});
export const centerStyle = b.styleDef({
    flexGrow: 1,
    height: constants.Height,
    minWidth: 0,
});

export const centeredWithinPageStyle = b.styleDef({
    position: "absolute",
    left: "50%",
    transform: "translate(-50%, 0)",
});
