// bobwai--example/src/controlPanelGenerator/copyToClipboard.ts
/** following code was copied from helpers--clipboard component */
export enum ContentProperty {
    innerHTML = "innerHTML",
    innerText = "innerText",
}

export function getShortcutKey(): string {
    return isMac() ? "Command" : "Ctrl";
}

export function isMac(): boolean {
    return navigator.platform === "MacIntel";
}

export function getSelectedText(): string {
    let text = "";

    if (typeof window.getSelection) {
        const sel = window.getSelection();
        if (sel !== null) {
            text = sel.toString();
        }
    }

    return text;
}

export function copyToClipboard(value: string, contentProperty: ContentProperty = ContentProperty.innerHTML) {
    const clipboardDiv = createClipboardDiv();
    document.body.appendChild(clipboardDiv);
    selectValueOnClipboardDiv(clipboardDiv, value, contentProperty);
    document.execCommand("copy");
    document.body.removeChild(clipboardDiv);
}

function createClipboardDiv() {
    const clipboardDiv = document.createElement("div");
    clipboardDiv.style.fontSize = "12pt";
    clipboardDiv.style.border = "0";
    clipboardDiv.style.padding = "0";
    clipboardDiv.style.margin = "0";
    clipboardDiv.style.position = "fixed";
    clipboardDiv.style["right"] = "-9999px";
    clipboardDiv.style.top = (window.pageYOffset || document.documentElement!.scrollTop) + "px";
    clipboardDiv.setAttribute("readonly", "");
    clipboardDiv.style.opacity = "0";
    clipboardDiv.style.pointerEvents = "none";
    /** this part was added just for using in example for copy spaces and tabs */
    clipboardDiv.style.whiteSpace = "pre";
    clipboardDiv.style.zIndex = "-1";
    clipboardDiv.setAttribute("tabindex", "0");
    clipboardDiv.innerHTML = "";
    return clipboardDiv;
}

function selectValueOnClipboardDiv(clipboardDiv: HTMLElement, value: string, contentProperty: ContentProperty) {
    clipboardDiv[contentProperty] = value;
    clipboardDiv.focus();

    let sel = window.getSelection();
    if (sel !== null) {
        sel.removeAllRanges();
    }

    const range = document.createRange();
    const firstChild = clipboardDiv.firstChild;
    const lastChild = clipboardDiv.lastChild;
    if (firstChild) {
        range.setStartBefore(firstChild);
    }
    if (lastChild) {
        range.setEndAfter(lastChild);
    }

    sel = window.getSelection();
    if (sel !== null) {
        sel.addRange(range);
    }
}
