// bobwai--chat/spec/GUI/store.ts
import * as b from "bobril";
import * as chat from "../../src/lib";
import * as Icon from "bobwai--icon";
import * as color from "bobwai--color";
import * as button from "bobwai--button";

export interface ITestStore {
    model: chat.IComment<number>[];
    lastId1: number;
    commentValue: string;
    activeCommentId: number;
    defaultRootCommentId: number;
    rootCommentAddArea: number;
    commentAddArea0: number;
    commentAddArea1: number;
}

let testStore: ITestStore = getDefaultTestStore();

export function getTestStore(): ITestStore {
    return testStore;
}

export function clearTestStore(): void {
    testStore = getDefaultTestStore();
}

function getDefaultTestStore(): ITestStore {
    return {
        model: [
            {
                id: 1,
                text: "This is comment.",
                userName: "John Doe",
                created: "01 December",
                icon: b.styledDiv(null, b.sprite(Icon.user_medium_png, color.Application)),
                replies: [
                    {
                        id: 12356,
                        text: "This is reply.",
                        userName: "John Doe",
                        created: "01 December",
                        icon: b.styledDiv(null, b.sprite(Icon.user_small_24_png, color.Application)),
                        replies: []
                    },
                    {
                        id: 7,
                        text: "This is reply. Michael Doe has his own icon.",
                        userName: "Michael Doe",
                        created: "01 December",
                        replies: [],
                        icon: b.styledDiv(null, b.sprite(Icon.user_small_24_png, color.Application))
                    }
                ]
            },
            {
                id: 2,
                text: "This is comment from anonymous user.",
                userName: "Anonymous",
                icon: b.styledDiv(null, b.sprite(Icon.user_medium_png, color.White)),
                created: "01 December",
                replies: []
            },
            {
                id: 3,
                text: "This is comment. Michael Doe has his own icon.",
                userName: "Michael Doe",
                icon: b.styledDiv(null, b.sprite(Icon.user_medium_png, color.Application)),
                created: "01 December",
                replies: []
            },
            {
                id: 4,
                text: "This is comment with extra control panel.",
                userName: "John Doe",
                icon: b.styledDiv(null, b.sprite(Icon.user_medium_png, color.Error)),
                created: "01 December",
                replies: [],
                controlPanel: button.create({
                    leftIcon: b.styledDiv(null, b.sprite(Icon.visible_small_png, color.Basic), { display: "inline-block" }),
                    leftIconHover: b.styledDiv(null, b.sprite(Icon.visible_small_png, color.Application), { display: "inline-block" }),
                    onClick: () => {
                        getTestStore().commentAddArea0++;
                    },
                    label: "VIEW AREA",
                    size: button.Size.Small,
                    color: button.Theme.None
                })
            },
            {
                id: 5,
                text: "This is selected comment.",
                userName: "John Doe",
                icon: b.styledDiv(null, b.sprite(Icon.user_medium_png, color.Error)),
                created: "01 December",
                replies: [],
                isSelected: true,
                isEditable: true
            },
            {
                id: 6,
                text: "This is selected comment with extra control panel.",
                userName: "John Doe",
                icon: b.styledDiv(null, b.sprite(Icon.user_medium_png, color.Error)),
                created: "01 December",
                replies: [],
                isSelected: true,
                controlPanel: button.create({
                    leftIcon: b.styledDiv(null, b.sprite(Icon.visible_small_png, color.Basic), { display: "inline-block" }),
                    leftIconHover: b.styledDiv(null, b.sprite(Icon.visible_small_png, color.Application), { display: "inline-block" }),
                    onClick: () => {
                        getTestStore().commentAddArea1++;
                    },
                    label: "VIEW AREA",
                    size: button.Size.Small,
                    color: button.Theme.None
                }),
                isEditable: true
            }
        ],
        lastId1: 8,
        commentValue: "",
        activeCommentId: 0,
        defaultRootCommentId: -1,
        rootCommentAddArea: 0,
        commentAddArea0: 0,
        commentAddArea1: 0
    };
}
