// bobwai--action-with-window/examples/src/actionContentHtmlElement.ts
import * as ActionWithWindow from "../../src/lib";
import * as b from "bobril";
import * as Helpers from "./helpers";
import * as Container from "bobwai--viewer-container";
import * as Example from "bobwai--example";
import * as Combobox from "bobwai--combobox";
import * as Button from "bobwai--button";
import { IBobrilNode } from "bobril";

let htmlElement: HTMLElement | undefined;
let isWindowVisible: boolean = false;
let preferredPlacement: ActionWithWindow.Placement | undefined;

export default function (): IBobrilNode[] {
    return [
        Container.create({
            isTest: true,
            header: "Action content is HTML element - primary for the using in II application (replacement for jquery picker)",
            description:
                "If action content is HTMLElement no action content is rendered, html element is used only for getting information about position",
            content: [
                getControlPanel(),
                getOpenWindowButton(),
                b.styledDiv("", mockSeparatorStyle),
                getMockHTMLElement(),
                b.styledDiv("", mockSeparatorStyle),
                htmlElement !== undefined && isWindowVisible && getActionWithWindow(),
            ],
        }),
    ];
}

function getControlPanel(): b.IBobrilNode {
    return Example.createControlPanel({
        items: [
            Example.createField({
                customValueContent: Example.createComboEnum({
                    comboboxFactory: Combobox.create,
                    enum: ActionWithWindow.Placement,
                    onChange: (v) => {
                        preferredPlacement = v;
                        b.invalidate();
                    },
                    value: preferredPlacement,
                }),
                label: "PreferredPlacement",
            }),
        ],
    });
}

function getOpenWindowButton(): b.IBobrilNode {
    return Button.create({
        label: "Open window",
        isDisabled: isWindowVisible,
        onClick: () => {
            isWindowVisible = true;
            b.invalidate();
        },
    });
}

function getMockHTMLElement(): b.IBobrilNode {
    return b.styledDiv(
        b.style(
            {
                tag: "span",
                component: {
                    postInitDom(_ctx: b.IBobrilCtx, _me: b.IBobrilCacheNode, element: HTMLElement) {
                        htmlElement = element;
                        b.invalidate();
                    },
                },
            },
            mockElementStyle
        ),
        mockCenterStyle
    );
}

function getActionWithWindow(): b.IBobrilNode {
    return ActionWithWindow.create({
        isVirtual: true,
        positioningMode: ActionWithWindow.PositioningMode.All,
        isWindowVisible: isWindowVisible,
        onCloseWindow: () => {
            isWindowVisible = false;
            b.invalidate();
        },
        actionContent: { relatedHTMLElement: htmlElement },
        windowContent: Helpers.getWindow(),
        maxWindowHeight: Helpers.windowSizeDefault,
        preferredPlacement: preferredPlacement,
    });
}

const mockSeparatorStyle = b.styleDef({
    height: 200,
});

const mockElementStyle = b.styleDef({ height: 20, background: "green", width: 20, display: "inline-block" });

const mockCenterStyle = b.styleDef({ textAlign: "center" });
