// bobwai--action-with-window/examples/src/fixedPosition.ts
import * as ActionWithWindow from "../../src/lib";
import * as Helpers from "./helpers";
import * as Container from "bobwai--viewer-container";
import { IBobrilNode } from "bobril";

export default function (): IBobrilNode[] {
    return [
        Container.create({
            header: "Window with fixed position",
            content: [
                Helpers.createVerticalSpace(),

                ActionWithWindow.create({
                    actionContent: Helpers.getButton(),
                    windowContent: Helpers.getWindow(),
                    fixPosition: true,
                    positioningMode: ActionWithWindow.PositioningMode.All,
                    preferredPlacement: ActionWithWindow.Placement.Right,
                    secondaryPreferredPlacement: ActionWithWindow.Placement.BottomRight,
                }),
            ],
            description: "When you set fixed position, window will not automatic calculate possible preferred placement and position.",
        }),
    ];
}
