// bobwai--action-with-window/examples/src/usageOfWidthWindow.ts
import * as ActionWithWindow from "../../src/lib";
import * as Helpers from "./helpers";
import * as Container from "bobwai--viewer-container";
import { IBobrilNode } from "bobril";

export default function (): IBobrilNode[] {
    return [
        Container.create({
            header: "Width of window same as actionContent, window has up/down icon used e.g for datepicker/timepicker and variable-input",
            content: [
                ActionWithWindow.create({
                    actionContent: Helpers.getButton(false, 400),
                    isFitWidth: true,
                    windowContent: Helpers.getWindowWithHeader({ autoWidth: true }),
                    hasArrow: true,
                    onUpdateSide: (isTop: boolean) => {
                        console.log("update side ", isTop);
                    },
                    hasShadow: true,
                    hasBorder: true,
                }),
            ],
        }),

        Container.create({
            header: "Width of window same as actionContent, window has up/down icon used e.g for datepicker/timepicker and variable-input",
            content: [
                ActionWithWindow.create({
                    actionContent: Helpers.getButton(),
                    isFitWidth: true,
                    windowContent: Helpers.getWindowWithHeader({
                        width: 400,
                    }),
                    hasArrow: true,
                    onUpdateSide: (isTop: boolean) => {
                        console.log("update side ", isTop);
                    },
                    hasShadow: true,
                    hasBorder: true,
                }),
            ],
        }),
    ];
}
