// bobwai--action-with-window/examples/src/visibleWindowByDefault.ts
import * as b from "bobril";
import * as ActionWithWindow from "../../src/lib";
import * as Helpers from "./helpers";
import * as Container from "bobwai--viewer-container";

let isWindowVisible: boolean = true;
export default function () {
    return Container.create({
        header: "Sample with overlay window visible by default",
        content: [
            b.styledDiv(
                "NOTE: In FIREFOX this window could be placed on wrong position" +
                    " (usually several pixels above correct position). " +
                    "This will happen only when page is loading fonts for the first time." +
                    " When you will refresh page (without clearing cache), this will not happen and possition " +
                    "will be ok. Please, don't report this as a bug because this will not happen in real application," +
                    " this is just bug of this example.",
                { color: "red", fontSize: "80%", paddingBottom: 10, maxWidth: 800 }
            ),
            ActionWithWindow.create({
                isWindowVisible: isWindowVisible,
                preferredPlacement: ActionWithWindow.Placement.BottomRight,
                secondaryPreferredPlacement: ActionWithWindow.Placement.TopRight,
                actionContent: Helpers.getButton(),
                windowContent: Helpers.getWindow(),
                onOpenWindow: () => {
                    isWindowVisible = true;
                    b.invalidate();
                },
                onCloseWindow: () => {
                    isWindowVisible = false;
                    b.invalidate();
                },
            }),
        ],
        description: "Click green button to toggle overlay window.",
    });
}
