// bobwai--action-with-window/examples/src/windowWithScrollbar.ts
import * as ActionWithWindow from "../../src/lib";
import * as Helpers from "./helpers";
import * as Container from "bobwai--viewer-container";
import { IBobrilNode } from "bobril";

export default function (): IBobrilNode[] {
    return [
        Container.create({
            header: "Sample with resize and scrollbar enabled and maxWindowHeight defined",
            content: [
                Helpers.createVerticalSpace(),

                ActionWithWindow.create({
                    actionContent: Helpers.getButton(),
                    windowContent: Helpers.getWindow(),
                    isResizeEnabled: true,
                    isScrollbarEnabled: true,
                    maxWindowHeight: Helpers.windowSizeDefault,
                }),
            ],
            description: "If you will resize window to even smaller size than is defined, it will still work correctly.",
        }),
        Container.create({
            header: "Sample with resize enabled but without scrollbar",
            content: [
                Helpers.createVerticalSpace(),

                ActionWithWindow.create({
                    actionContent: Helpers.getButton(),
                    windowContent: Helpers.getWindow(),
                    isResizeEnabled: true,
                }),
            ],
            description:
                "To see how it works, get it to upper side with not enough space" +
                " - you should still the upper part of window (with label)." +
                " This could be used in combination with content which has inner scrollbar.",
        }),
    ];
}
