// bobwai--action-with-window/examples/src/windowWithOverflow.ts
import * as b from "bobril";
import * as ActionWithWindow from "../../src/lib";
import * as Helpers from "./helpers";
import * as Container from "bobwai--viewer-container";

const longText = b.styledDiv("This is overlay window containing long text going outside of it.", {
    position: "absolute",
    whiteSpace: "nowrap",
    border: "1px solid skyblue",
});

export default function () {
    return [
        Container.create({
            header: "Overflow sample: default",
            content: [
                ActionWithWindow.create({
                    actionContent: Helpers.getButton(),
                    windowContent: Helpers.getWindow(longText),
                }),
            ],
            description: "Overflowed content is hidden.",
        }),

        Container.create({
            header: "Overflow sample: isOverflowVisible = true",
            content: [
                ActionWithWindow.create({
                    actionContent: Helpers.getButton(),
                    windowContent: Helpers.getWindow(longText),
                    isOverflowVisible: true,
                }),
            ],
            description: "Overflowed content is visible. This is useful when component uses custom shadow etc.",
        }),
    ];
}
