// bobwai--action-with-window/examples/src/windowWithCloseButton.ts
import * as b from "bobril";
import * as ActionWithWindow from "../../src/lib";
import * as Helpers from "./helpers";
import * as Container from "bobwai--viewer-container";
import * as Button from "bobwai--button";

let isWindowVisible: boolean = false;

const closeButton = Button.create({
    label: "Close",
    onClick: () => {
        isWindowVisible = false;
        b.invalidate();
    },
});

export default function () {
    return Container.create({
        header: "Sample with close button inside overlay window",
        content: [
            ActionWithWindow.create({
                isWindowVisible: isWindowVisible,
                actionContent: Helpers.getButton(),
                windowContent: Helpers.getWindow(closeButton),
                onOpenWindow: () => {
                    isWindowVisible = true;
                    b.invalidate();
                },
                onCloseWindow: () => {
                    isWindowVisible = false;
                    b.invalidate();
                },
            }),
        ],
        description: "Click green button to toggle overlay window. Click close button to hide overlay window.",
    });
}
