// bobwai--action-with-window/examples/src/windowContentChanging.ts
import * as b from "bobril";
import * as ActionWithWindow from "../../src/lib";
import * as Helpers from "./helpers";
import * as Container from "bobwai--viewer-container";
import * as Button from "bobwai--button";

let windowHeight = Helpers.windowSizeDefault;

export default function () {
    return Container.create({
        header: "Sample with height of window content change",
        content: [
            ActionWithWindow.create({
                actionContent: Helpers.getButton(),
                windowContent: Helpers.getWindow(
                    Button.create({
                        label: "Update height",
                        onClick: () => {
                            windowHeight = 600;
                            b.invalidate();
                        },
                    }),
                    false,
                    200,
                    windowHeight
                ),
                preferredPlacement: ActionWithWindow.Placement.BottomRight,
                secondaryPreferredPlacement: ActionWithWindow.Placement.TopRight,
                positioningMode: ActionWithWindow.PositioningMode.All,
            }),
        ],
        description: "Click green button to toggle overlay window. Click close button to hide overlay window.",
    });
}
