// bobwai--action-with-window/examples/src/windowWithResizeAndScrollbar.ts
import * as ActionWithWindow from "../../src/lib";
import * as Helpers from "./helpers";
import * as Container from "bobwai--viewer-container";
import { IBobrilNode } from "bobril";

export default function (): IBobrilNode[] {
    return [
        Container.create({
            header: "Sample with resize & scrollbar enabled",
            content: [
                ActionWithWindow.create({
                    actionContent: Helpers.getButton(),
                    windowContent: Helpers.getWindow(),
                    isResizeEnabled: true,
                    isScrollbarEnabled: true,
                }),
            ],
            description: "You have to resize window to have smaller space than window needs to use.",
        }),
        Container.create({
            header: "Sample with resize & scrollbar enabled (on left side)",
            content: [
                Helpers.createVerticalSpace(),

                ActionWithWindow.create({
                    actionContent: Helpers.getButton(),
                    windowContent: Helpers.getWindow(),
                    positioningMode: ActionWithWindow.PositioningMode.TopBottom,
                    preferredPlacement: ActionWithWindow.Placement.BottomLeft,
                    secondaryPreferredPlacement: ActionWithWindow.Placement.TopLeft,
                    isResizeEnabled: true,
                    isScrollbarEnabled: true,
                }),
            ],
            description: "You have to resize window to have smaller space than window needs to use.",
        }),
    ];
}
