// bobwai--l-app/src/lib.ts
import * as b from "bobril";
import * as styles from "./styles";
import * as c from "./publicConstants";

export const create = b.createComponent<IData>({
    id: "bobwai--l-app",
    render(ctx: IContext, me: b.IBobrilNode): void {
        b.style(me, styles.baseStyle);
        if (!ctx.data.isMobile) {
            me.style = {
                minWidth: ctx.data.minWidth ? ctx.data.minWidth : c.MinWidth,
            };
        }
        me.children = [
            ctx.data.children,
            ctx.data.appHeaderContent,
            ctx.data.appToolbarContent,
            ctx.data.appBreadcrumbContent,
            ctx.data.appSidebarContent,
            ctx.data.appSideNotificationContent,
        ];
    },
});

export interface IData {
    children?: b.IBobrilChildren;
    appHeaderContent?: b.IBobrilChildren;
    appBreadcrumbContent?: b.IBobrilChildren;
    appToolbarContent?: b.IBobrilChildren;
    appSidebarContent?: b.IBobrilChildren;
    appSideNotificationContent?: b.IBobrilChildren;
    isMobile?: boolean;
    minWidth?: number;
}

export * from "./publicConstants";

interface IContext extends b.IBobrilCtx {
    data: IData;
}

// added by script - BEGIN
export const LApp = create;
// added by script - END
