// bobwai--toolbar/src/styles.ts
import * as b from "bobril";
import * as color from "bobwai--color";
import * as c from "./publicConstants";

export const bothSidePadding = 10;
export const expandButtonWidth = 25;
export const borderBottomWidth = 1;
export const centerContentMinWidth = 50;

export const toolbarStyle = b.styleDef({
    width: "100%",
    height: c.Height,
    overflow: "auto",
    position: "relative",
    boxSizing: "border-box",
    padding: `0 ${bothSidePadding}px`,
    background: color.Grey100,
    borderBottomWidth: borderBottomWidth,
    borderBottomStyle: "solid",
    borderBottomColor: color.Grey400
});

export const toolbarSmallStyle = b.styleDefEx(toolbarStyle, {
    height: c.SmallHeight
});

export const itemsLeftDivStyle = b.styleDef({
    cssFloat: "left",
    height: "100%"
});

export const itemsRightDivStyle = b.styleDef({
    height: "100%",
    display: "flex",
    position: "absolute",
    right: bothSidePadding
});

export const toolbarItem = b.styleDef({
    cssFloat: "left",
    height: "100%"
});

export const centerContentStyle = b.styleDef({
    height: "100%",
    overflow: "hidden",
    position: "absolute",
    minWidth: `${centerContentMinWidth}px`
});

export const rightExpandButton = b.styleDef({
    position: "absolute",
    right: 0,
    height: "100%",
    width: `${expandButtonWidth}px`,
    borderLeftWidth: 1,
    borderLeftColor: color.Grey400,
    borderLeftStyle: "solid",
    textAlign: "center",
    backgroundColor: color.Grey100
});

export const rightExpandIcon = b.styleDef({
    verticalAlign: "middle",
    lineHeight: `${c.Height - borderBottomWidth}px`
});

export const rightExpandContent = b.styleDef({
    height: c.Height,
    border: "solid",
    borderWidth: "1px",
    borderColor: color.Grey400,
    boxShadow: "0px 1px 3px 0px rgba(0, 0, 0, 0.2)",
    position: "relative",
    right: "0px"
});

export const rightExpandItem = b.styleDef({
    cssFloat: "left",
    height: "100%",
    backgroundColor: color.White
});
