// bobwai--context-menu/examples/helpers/wrappers.ts
import * as b from "bobril";
import * as Fileupload from "helpers--fileupload";

export function linkWrapper(content: b.IBobrilNode): b.IBobrilNode {
    return {
        tag: "a",
        attrs: { href: "https://www.google.com", target: "_blank" },
        children: content
    };
}

export function fileUploadWrapper(content: b.IBobrilChild): b.IBobrilNode {
    return b.withKey(
        Fileupload.Area.create({
            children: content,
            localizer: {
                fileSizeInvalid: (maxFileSize: number) => {
                    return "Invalid Size " + maxFileSize;
                },
                fileTypeInvalid: (_a: string) => {
                    return "Invalid file type";
                },
                invalidResponseFromServer: () => {
                    return "Invalid response from server";
                }
            },
            options: { method: "POST", url: "test" },
            uploader: Fileupload.Uploader.registerUploader(() => {
                return;
            })
        }),
        "Key"
    );
}
