// bobwai--color/src/dynamicColors.ts
import * as b from "bobril";
import * as constants from "./publicConstants";
import { getTransparentColor } from "./helper";

// Each application and sample company can have different

// ---------- Application color -----------
export const DefaultApplicationColor = "#72bf44"; // 'rgba(114, 191, 78, 1)'
let currentApplicationColor = DefaultApplicationColor;

export function Application(): string {
    return currentApplicationColor;
}

export function Application_10(): string {
    return getTransparentColor(currentApplicationColor, 0.1);
}

export function Application_30(): string {
    return getTransparentColor(currentApplicationColor, 0.3);
}

export function Application_40(): string {
    return getTransparentColor(currentApplicationColor, 0.4);
}

export function Application_64(): string {
    return getTransparentColor(currentApplicationColor, 0.64);
}

export function setApplicationColor(color: string): void {
    currentApplicationColor = color;
    b.invalidateStyles();
}

export function restoreDefaultApplicationColor(): void {
    setApplicationColor(DefaultApplicationColor);
}

// ---------- Basic color -----------
export const DefaultBasicColor = "#40484b"; // 'rgba(64, 72, 75, 1)'
let currentBasicColor = DefaultBasicColor;

export function Basic(): string {
    return currentBasicColor;
}

export function Basic_08(): string {
    return getTransparentColor(currentBasicColor, 0.08);
}

/** Deprecated - use Basic_Opacity200 instead of this */
export function Basic_16(): string {
    return getTransparentColor(currentBasicColor, 0.16);
}

/** Deprecated - use Basic_Opacity300 instead of this */
export function Basic_24(): string {
    return getTransparentColor(currentBasicColor, 0.24);
}

export function Basic_64(): string {
    return getTransparentColor(currentBasicColor, 0.64);
}

export function Basic_30(): string {
    return getTransparentColor(currentBasicColor, 0.3);
}

export function Basic_32(): string {
    return getTransparentColor(currentBasicColor, 0.32);
}

export function Basic_80(): string {
    return getTransparentColor(currentBasicColor, 0.8);
}

export function setBasicColor(color: string): void {
    currentBasicColor = color;
    b.invalidateStyles();
}

export function restoreDefaultBasicColor(): void {
    setBasicColor(DefaultBasicColor);
}

// ---------- Primary color -----------
export const DefaultPrimaryColor = "#004a65"; // 'rgba(0, 74, 101, 1)'
let currentPrimaryColor = DefaultPrimaryColor;

export function Primary(): string {
    return currentPrimaryColor;
}

export function setPrimaryColor(color: string): void {
    currentPrimaryColor = color;
    b.invalidateStyles();
}

export function restoreDefaultPrimaryColor(): void {
    setPrimaryColor(DefaultPrimaryColor);
}

// ---------- Secondary color -----------
export const DefaultSecondaryColor = "#8bb81d"; // 'rgba(139, 184, 29, 1)'
let currentSecondaryColor = DefaultSecondaryColor;

export function Secondary(): string {
    return currentSecondaryColor;
}

export function setSecondaryColor(color: string): void {
    currentSecondaryColor = color;
    b.invalidateStyles();
}

export function restoreDefaultSecondaryColor(): void {
    setSecondaryColor(DefaultSecondaryColor);
}

// ---------- Chart colors - list sorting made by UX: 500/300/100/700 -----------
const chartColorsSortedList = [
    constants.Chart_Green500,
    constants.Chart_Blue500,
    constants.Chart_Violet500,
    constants.Chart_Yellow500,
    constants.Chart_Brown500,
    constants.Chart_Purple500,
    constants.Chart_Orange500,
    constants.Chart_GreenB500,
    constants.Chart_Green300,
    constants.Chart_Blue300,
    constants.Chart_Violet300,
    constants.Chart_Yellow300,
    constants.Chart_Brown300,
    constants.Chart_Purple300,
    constants.Chart_Orange300,
    constants.Chart_GreenB300,
    constants.Chart_Green100,
    constants.Chart_Blue100,
    constants.Chart_Violet100,
    constants.Chart_Yellow100,
    constants.Chart_Brown100,
    constants.Chart_Purple100,
    constants.Chart_Orange100,
    constants.Chart_GreenB100,
    constants.Chart_Yellow700,
    constants.Chart_Orange700,
    constants.Chart_GreenB700
];

export function getChartItemColor(index: number): string {
    if (index < chartColorsSortedList.length) {
        return chartColorsSortedList[index];
    }

    return chartColorsSortedList[index % chartColorsSortedList.length];
}
