// bobwai--color/src/publicConstants.ts
// Application specific colors moved to "dynamic form", see dynamicColors.ts

// Notifications
export const Success = "#69b027"; // 'rgba(105, 176, 39, 1)';
export const Warning = "#f28918"; // 'rgba(242, 137, 24, 1)'
export const Error = "#e34243"; // 'rgba(227, 66, 67, 1)'
export const Info = "#3598db"; // 'rgba(53, 152, 219, 1)'

// Greys
export const White = "#ffffff"; // 'rgba(255, 255, 255, 1)'
export const Grey000 = "#fcfcfc"; // 'rgba(252, 252, 252, 1)';
export const Grey100 = "#f5f5f5"; // 'rgba(245, 245, 245, 1)'
export const Grey200 = "#f0f0f0"; // 'rgba(240, 240, 240, 1)'
export const Grey250 = "#e2e2e2"; // 'rgba(226, 226, 226, 1)'
export const Grey300 = "#d4d4d4"; // rgba(212, 212, 212, 1)
export const Grey400 = "#cccccc"; // 'rgba(204, 204, 204, 1)'
export const Grey500 = "#aeaeae"; // 'rgba(174, 174, 174, 1)'
export const Grey600 = "#818181"; //  rgba(129, 129, 129, 1)
export const Black = "#000000"; // 'rgba(0, 0, 0, 1)'

// With transparency
export const White_08 = "rgba(255, 255, 255, 0.08)"; // #ffffff
export const White_16 = "rgba(255, 255, 255, 0.16)"; // #ffffff
export const White_24 = "rgba(255, 255, 255, 0.24)"; // #ffffff
export const White_30 = "rgba(255, 255, 255, 0.3)"; // #ffffff
export const White_50 = "rgba(255, 255, 255, 0.5)"; // #ffffff
export const White_60 = "rgba(255, 255, 255, 0.6)"; // #ffffff˝
export const White_64 = "rgba(255, 255, 255, 0.64)"; // #ffffff˝
export const White_75 = "rgba(255, 255, 255, 0.75)"; // #ffffff
export const White_80 = "rgba(255, 255, 255, 0.80)"; // #ffffff
export const Success_20 = "rgba(105, 176, 39, 0.2)"; // #69b027
export const Success_30 = "rgba(105, 176, 39, 0.3)"; // #69b027
export const Error_05 = "rgba(227, 66, 67, 0.05)"; // #e34243
export const Error_10 = "rgba(227, 66, 67, 0.1)"; // #e34243
export const Error_20 = "rgba(227, 66, 67, 0.2)"; // #e34243
export const Error_30 = "rgba(227, 66, 67, 0.3)"; // #e34243
export const Error_32 = "rgba(201, 43, 0, 0.32)"; // #c92b00
export const Error_40 = "rgba(227, 66, 67, 0.4)"; // #e34243
export const Warning_05 = "rgba(242, 137, 24, 0.05)"; // #f28918
export const Warning_10 = "rgba(242, 137, 24, 0.1)"; // #f28918
export const Warning_30 = "rgba(242, 137, 24, 0.3)"; // #f28918
export const Info_30 = "rgba(53, 152, 219, 0.3)"; // #f7c6c6
export const Black_16 = "rgba(0, 0, 0, 0.16)"; // #000000
export const Black_20 = "rgba(0, 0, 0, 0.2)"; // #000000
export const Black_24 = "rgba(0, 0, 0, 0.24)"; // #000000
export const Black_42 = "rgba(0, 0, 0, 0.42)"; // #000000 (overlays)
export const Black_75 = "rgba(0, 0, 0, 0.75)"; // #000000
export const Grey300_30 = "rgba(212, 212, 212, 0.3)"; // #d4d4d4
export const Grey500_15 = "rgba(174, 174, 174, 0.15)"; // #aeaeae
export const Grey500_30 = "rgba(174, 174, 174, 0.3)"; // #aeaeae
export const Grey500_80 = "rgba(174, 174, 174, 0.8)"; // #aeaeae
export const Grey600_30 = "rgba(129, 129, 129, 0.3)"; // #818181

// Invisible
export const Transparent = "rgba(255, 255, 255, 0)"; // #ffffff

// Additional colors
export const Error300 = "#bd3737";
export const Error500 = "#962c2c";
export const Error_05_solid = "#fef6f6";
export const Warning300 = "#cc7314";
export const Warning500 = "#a65e11";
export const Warning_05_solid = "#fef9f3";
export const Info_05_solid = "#f5fafd";

// Mostly used for charts
export const Chart_Green_32 = "rgba(18, 179, 154, 0.32)"; // #12b39a
export const Chart_Green100 = "#a1e6da";
export const Chart_Green300 = "#52ccba";
export const Chart_Green500 = "#12b39a";
export const Chart_Blue100 = "#a1cfe6";
export const Chart_Blue300 = "#52a3cc";
export const Chart_Blue500 = "#127db2";
export const Chart_Violet_32 = "rgba(126, 87, 194, 0.32)"; // #7e57c2
export const Chart_Violet100 = "#d1c4e9";
export const Chart_Violet300 = "#9575cd";
export const Chart_Violet500 = "#7e57c2";
export const Chart_Yellow_32 = "rgba(224, 193, 62, 0.32)"; // #e0c13e
export const Chart_Yellow100 = "#fdefba";
export const Chart_Yellow300 = "#fce68d";
export const Chart_Yellow500 = "#fadd60";
export const Chart_Yellow700 = "#e0c13e";
export const Chart_Brown100 = "#d7ccc8";
export const Chart_Brown300 = "#bcaaa4";
export const Chart_Brown500 = "#8d6e63";
export const Chart_Purple100 = "#f7d5ef";
export const Chart_Purple300 = "#d97cc1";
export const Chart_Purple500 = "#cc0099";
export const Chart_Orange100 = "#f2d2b6";
export const Chart_Orange300 = "#f2ac61";
export const Chart_Orange500 = "#f28918";
export const Chart_Orange700 = "#c66b29";
export const Chart_GreenB100 = "#d0e5b1";
export const Chart_GreenB300 = "#aecc85";
export const Chart_GreenB500 = "#69b027";
export const Chart_GreenB700 = "#568a3e";
