// bobwai--l-view-content-with-footer/src/lib.ts
import * as b from "bobril";
import * as c from "./publicConstant";

export let create = b.createComponent<IData>({
    id: "bobwai--l-view-content-with-footer",
    render(ctx: IContext, me: b.IBobrilNode): void {
        b.style(me, layoutStyle, ctx.data.customFooterHeight && { paddingBottom: ctx.data.customFooterHeight });
        me.children = [
            b.styledDiv(ctx.data.bodyContent, contentStyle),
            b.styledDiv(ctx.data.footerContent, footerStyle, ctx.data.customFooterHeight && { height: ctx.data.customFooterHeight })
        ];
    }
});

export interface IData {
    bodyContent: b.IBobrilChildren;
    footerContent: b.IBobrilChildren;
    customFooterHeight?: number;
}

export * from "./publicConstant";

interface IContext extends b.IBobrilCtx {
    data: IData;
}

const layoutStyle = b.styleDef(
    {
        boxSizing: "border-box",
        height: "100%",
        paddingBottom: c.FooterHeight
    },
    {},
    "l-view-content-with-footer"
);

const contentStyle = b.styleDef(
    [
        {
            minHeight: "100%",
            width: "100%",
            cssFloat: "left"
        }
    ],
    {},
    "l-view-content-with-footer-body"
);

const footerStyle = b.styleDef(
    [
        {
            width: "100%",
            height: c.FooterHeight,
            cssFloat: "left"
        }
    ],
    {},
    "l-view-content-with-footer-footer"
);

// added by script - BEGIN
export const LViewContentWithFooter = create;
// added by script - END
