// bobwai--animation/src/fadeInFadeOut.ts
import * as b from "bobril";
import { IShowHideAnimation, IAnimationTransition } from "./lib";

export function fadeInFadeOutAnimation(): IShowHideAnimation {
    return {
        baseStyle: fadeInFadeOutBaseStyle,
        showStyle: fadeInFadeOutShowStyle,
        hideStyle: fadeInFadeOutHideStyle,
        showStyleDuration: localVariables.fadeInFadeOutDuration,
        hideStyleDuration: localVariables.fadeInFadeOutDuration
    };
}

export function fadeInFadeOutTransition(): IAnimationTransition {
    return {
        defaultTransition: defaultTransactionStyle,
        returnTransition: defaultTransactionStyle
    };
}

namespace localVariables {
    export const fadeInFadeOutDuration: number = 200;
}

const defaultTransactionStyle: b.IBobrilStyleDef = b.styleDef({
    transition: `all ${localVariables.fadeInFadeOutDuration}ms ease-in-out`
});

const fadeInFadeOutBaseStyle: b.IBobrilStyleDef = b.styleDef({
    opacity: 0
});

const fadeInFadeOutShowStyle: b.IBobrilStyleDef = b.styleDef([
    defaultTransactionStyle,
    {
        opacity: 1
    }
]);

const fadeInFadeOutHideStyle: b.IBobrilStyleDef = b.styleDef({
    opacity: 0
});
