// bobwai--context-menu/src/styles.ts
import * as b from "bobril";
import * as Color from "bobwai--color";
import * as Font from "bobwai--font";
import * as Mixins from "bobwai--style-mixins";
import * as Constants from "./publicConstants";

export const shortcutRight = 20;
const shortcutTop = 0;
const itemHeight = 40;
const itemBorder = `1px solid ${Color.Grey400}`;
const itemWithEmptyIconPaddingLeft = 40;
const itemIconDisabledOpacity = 0.3;
const itemWithIconGap = 10;
const iconContentMarginRight = 10;
const itemWithSubMenuPaddingRight = 28;
const subMenuIconRight = 5;
const subMenuIconTop = 0;
const itemWithSeparatorBorder = `1px solid ${Color.Grey300}`;
const contextMenuZIndex = 1;

export const contextMenuStyle = b.styleDef(
    {
        boxShadow: `0px 1px 3px ${Color.Black_20}`,
        position: "fixed",
        minWidth: Constants.MinWidth,
        backgroundColor: Color.White,
        boxSizing: "border-box",
        zIndex: contextMenuZIndex,
        borderTop: itemBorder,
        borderBottom: itemBorder,
        outline: "none"
    },
    {},
    "bobwai--context-menu"
);

export const contextMenuStaticStyle = b.styleDefEx(
    contextMenuStyle,
    {
        position: "static"
    },
    {},
    "bobwai--context-menu-static"
);

export const itemStyle = b.styleDef(
    [
        Mixins.centerVerticalParent(itemHeight),
        Mixins.getClickableElement(),
        {
            outline: "0 !important",
            lineHeight: `${itemHeight}px`,
            paddingLeft: Constants.ItemPaddingLeft,
            paddingRight: Constants.ItemPaddingLeft,
            position: "relative",
            boxSizing: "border-box",
            borderLeft: itemBorder,
            borderRight: itemBorder,
            color: Color.Basic,
            background: Constants.ItemBackgroundColor,
            whiteSpace: "nowrap"
        }
    ],
    {},
    "bobwai--context-menu-item"
);

export const itemHoverStyle = b.styleDefEx(
    itemStyle,
    {
        backgroundColor: Constants.ItemHoverColor
    },
    {},
    "is-hover"
);

export const itemContentStyle = b.styleDef(Mixins.centerVerticalChildInlineBlock(), {}, "bobwai--context-menu-item-content");

export const itemWithIconStyle = b.styleDef(
    {
        paddingLeft: itemWithIconGap
    },
    {},
    "bobwai--context-menu-item-icon"
);

export const itemWithEmptyIconStyle = b.styleDefEx(itemStyle, {
    paddingLeft: itemWithEmptyIconPaddingLeft
});

export const itemIconDisabledStyle = b.styleDef({
    opacity: itemIconDisabledOpacity
});

export const disabledStyle = b.styleDefEx(
    itemStyle,
    {
        cursor: "default",
        color: Color.Grey500
    },
    { hover: { backgroundColor: Constants.ItemBackgroundColor } },
    "is-disabled"
);

export const iconContentStyle = b.styleDef(
    [
        Mixins.centerVerticalChildInlineBlock(),
        {
            marginRight: iconContentMarginRight,
            lineHeight: 0,
            verticalAlign: "middle"
        }
    ],
    {},
    "bobwai--context-menu-icon-content"
);

export const itemWithSubMenuStyle = b.styleDef(
    {
        position: "relative",
        paddingRight: itemWithSubMenuPaddingRight
    },
    {},
    "bobwai--context-menu-item-submenu"
);

export const subMenuIconStyle = b.styleDef(
    {
        position: "absolute",
        right: subMenuIconRight,
        top: subMenuIconTop
    },
    {},
    "bobwai--context-menu-item-submenu-icon"
);

export const itemTitleStyle = b.styleDef(
    {
        color: Color.Grey500
    },
    { hover: { background: Constants.ItemBackgroundColor } },
    "context-menu-item-title"
);

export const itemWithSeparatorStyle = b.styleDef(
    {
        borderBottom: itemWithSeparatorBorder,
        cursor: "default"
    },
    {},
    "bobwai--context-menu-item-separator"
);

export const titleWithoutPaddingStyle = b.styleDefEx(
    itemWithEmptyIconStyle,
    {
        paddingLeft: iconContentMarginRight
    },
    {},
    "bobwai--context-menu-title-without-padding"
);

export const itemSelectedStyle = b.styleDefEx(
    itemStyle,
    {
        background: Color.Application_30
    },
    {},
    "bobwai--context-menu-item-selected"
);

export const shortcutStyle = b.styleDef(
    [
        Mixins.centerVerticalChildInlineBlock(),
        {
            fontWeight: Font.FontWeightRegular,
            color: Color.Grey500,
            display: "inline-block",
            position: "absolute",
            right: shortcutRight,
            top: shortcutTop
        }
    ],
    {},
    "bobwai--context-menu-item-shortcut"
);
