// bobwai--l-block/src/lib.ts
import * as b from "bobril";

export const create = b.createComponent<IData>({
    id: "bobwai--l-block",
    render(ctx: IContext, me: b.IBobrilNode): void {
        me.children = ctx.data.children;
        b.style(me, {
            paddingLeft: ctx.data.paddingLeft,
            paddingTop: ctx.data.paddingTop,
            paddingRight: ctx.data.paddingRight,
            paddingBottom: ctx.data.paddingBottom,
            width: ctx.data.width,
            minWidth: ctx.data.minWidth,
            height: ctx.data.height,
            background: ctx.data.background,
        });
    },
});

export interface IData {
    children: b.IBobrilChildren;
    paddingLeft?: number;
    paddingTop?: number;
    paddingRight?: number;
    paddingBottom?: number;
    width?: number | string;
    minWidth?: number | string;
    height?: number | string;
    background?: string;
}

interface IContext extends b.IBobrilCtx {
    data: IData;
}

// added by script - BEGIN
export const LBlock = create;
// added by script - END
