// bobwai--textarea/spec/GUI/layout.ts
import * as b from "bobril";
import { Textarea } from "../../src/lib";
import { getTestStore } from "./store";

export function createLayout(): void {
    b.init(() => {
        return [
            Textarea({
                value: getTestStore().value,
                onChange: (v) => {
                    getTestStore().value = v;
                    b.invalidate();
                },
            }),
            Textarea({
                value: getTestStore().value,
                onChange: (v) => {
                    getTestStore().value = v;
                    b.invalidate();
                },
                isDisabled: true,
                rows: 6,
            }),
            Textarea({
                value: getTestStore().value,
                onChange: (v) => {
                    getTestStore().value = v;
                    b.invalidate();
                },
                maxLength: 10, // bot ignore it
                rows: 3,
            }),
            Textarea({
                value: getTestStore().value,
                onChange: (v) => {
                    getTestStore().value = v;
                    b.invalidate();
                },
                isReadOnly: true,
            }),
            Textarea({
                value: getTestStore().value,
                onChange: (v) => {
                    getTestStore().value = v;
                    b.invalidate();
                },
                onFocusIn: () => {
                    getTestStore().isFocused = true;
                },
                onFocusOut: () => {
                    getTestStore().isFocused = false;
                },
            }),
        ];
    });
}
