// bobwai--l-view-login/src/styles.ts
import * as b from "bobril";
import * as c from "./publicConstants";

export const loginStyle = b.styleDef(
    {
        height: "100%",
        width: "100%;",
        textAlign: "center",
    },
    {},
    "l-view-login"
);

export const loginFirstStyle = b.styleDef(
    {
        cssFloat: "left",
        width: `${c.borderPercent}%`,
        height: "100%",
        position: "relative",
    },
    {},
    "l-view-login-first"
);

export const loginFirstMobileStyle = b.styleDefEx(
    loginFirstStyle,
    {
        cssFloat: "none",
        height: `${c.borderPercent}%`,
        width: "100%",
    },
    {},
    "l-view-login-first-mobile"
);

export const loginSecondStyle = b.styleDef(
    {
        position: "relative",
        cssFloat: "right",
        width: `${100 - c.borderPercent}%`,
        height: "100%",
        overflow: "auto",
    },
    {},
    "l-view-login-second"
);

export const loginSecondMobileStyle = b.styleDefEx(
    loginSecondStyle,
    {
        width: "100%",
        height: `${100 - c.borderPercent}%`,
        overflow: "visible",
    },
    {},
    "l-view-login-second-mobile"
);

export const logoStyle = b.styleDef(
    {
        position: "absolute",
        top: "50%",
        left: "50%",
        transform: "translateX(-50%) translateY(-50%)",
    },
    {},
    "l-view-login-logo"
);

export const logoInnerStyle = b.styleDef(
    {
        display: "inline-block",
    },
    {},
    "l-view-login-logo-inner"
);

export const contentTableStyle = b.styleDef(
    {
        display: "table",
        width: "100%",
        height: "100%",
    },
    {},
    "l-view-login-table"
);

export const contentWithFooterStyle = b.styleDefEx(contentTableStyle, {
    display: "flex",
    justifyContent: "center",
    alignItems: "center",
    width: "auto",
});

export const contentCellStyle = b.styleDef(
    {
        display: "table-cell",
        textAlign: "center",
        verticalAlign: "middle",
    },
    {},
    "l-view-login-content-cell"
);

export const contentInnerStyle = b.styleDef(
    {
        display: "inline-block",
    },
    {},
    "l-view-login-content-inner"
);

export const contentInnerMobileStyle = b.styleDefEx(
    contentInnerStyle,
    {
        display: "block",
    },
    {},
    "l-view-login-content-inner-mobile"
);

export const backLinkContentStyle = b.styleDef(
    {
        position: "absolute",
        top: 18,
        right: 24,
    },
    {},
    "l-view-login-back-link"
);

export const backLinkContentMobileStyle = b.styleDefEx(backLinkContentStyle, {
    position: "relative",
    top: 24,
    right: 50,
    textAlign: "right",
});
