// bobwai--composition--sidebar/src/wizard.ts
import * as b from "bobril";
import * as FormElements from "bobwai--form-elements";
import * as SidebarItem from "bobwai--sidebar-item";
import * as SidebarItemCollapsible from "bobwai--sidebar-item-collapsible";
import * as dataHelper from "../src/dataHelper";

export function getItems(): b.IBobrilNode[] {
    return dataHelper.getItemsFromData(rootData);
}

const mockContent = b.styledDiv("some content", { backgroundColor: "grey", height: 80 });

const rootData: SidebarItemCollapsible.IData[] = [
    {
        id: "1",
        onClick: onClick,
        title: "Attributes",
        isHighlighted: true,
        isCollapsible: true,
        collapsibleContent: mockContent,
        iconContent: dataHelper.getIconAttributes(),
        actions: [dataHelper.getStateIcon()],
        actionsVisibleOnlyOnHover: true,
    },
    {
        id: "2",
        onClick: onClick,
        title: "Variables",
        isHighlighted: true,
        isCollapsible: true,
        collapsibleContent: mockContent,
        iconContentError: dataHelper.getIconVariablesError(),
        validationState: FormElements.ValidationState.Error,
    },
    {
        id: "3",
        onClick: onClick,
        title: "Language",
        isHighlighted: true,
        isCollapsible: true,
        collapsibleContent: mockContent,
        iconContentWarning: dataHelper.getIconLanguageWarning(),
        validationState: FormElements.ValidationState.Warning,
    },
    {
        id: "4",
        onClick: onClick,
        title: "Dynamic zones",
        isHighlighted: true,
        isCollapsible: true,
        collapsibleContent: mockContent,
        iconContent: dataHelper.getIconDynamicZones(),
    },
];

function onClick(item: SidebarItem.IData): boolean {
    item.isExpanded = !item.isExpanded;
    b.invalidate();
    return true;
}
