// bobwai--sidebar-footer/src/lib.ts
import * as b from "bobril";
import * as Color from "bobwai--color";

export let create = b.createComponent<IData>({
    id: "bobwai--sidebar-footer",
    render(ctx: IContext, me: b.IBobrilNode): void {
        me.children = [
            ctx.data.leftContent && b.styledDiv(ctx.data.leftContent, contentStyle, leftContentStyle),
            ctx.data.leftContent && ctx.data.rightContent && b.styledDiv("", contentStyle, separatorStyle),
            ctx.data.rightContent && b.styledDiv(ctx.data.rightContent, contentStyle, rightContentStyle)
        ];
        b.style(me, sidebarFooterStyle);
    }
});

export interface IData {
    leftContent?: b.IBobrilChildren;
    rightContent?: b.IBobrilChildren;
}

interface IContext extends b.IBobrilCtx {
    data: IData;
}

namespace localVariables {
    export const paddingLeft = 25;
    export const seperatorPadding = 12;
    export const paddingBottom = 10;
    export const paddingTop = 10;
    export const separatorBackground = Color.Grey300;
    export const separatorHeight = 20;
    export const separatorWidth = 1;
    export const fontSize = 12;
    export const fontColor = Color.Grey500;
    export const leftContentBottomMargin = 4;
}

const sidebarFooterStyle = b.styleDef({
    padding: `${localVariables.paddingTop}px 0 ${localVariables.paddingBottom}px ${localVariables.paddingLeft}px`,
    whiteSpace: "nowrap"
});

const contentStyle = b.styleDef({
    display: "inline-block",
    verticalAlign: "bottom"
});

const separatorStyle = b.styleDefEx(contentStyle, {
    width: localVariables.separatorWidth,
    backgroundColor: localVariables.separatorBackground,
    marginLeft: localVariables.seperatorPadding,
    marginRight: localVariables.seperatorPadding,
    height: localVariables.separatorHeight
});

const rightContentStyle = b.styleDefEx(contentStyle, {
    fontSize: localVariables.fontSize,
    color: localVariables.fontColor
});

const leftContentStyle = b.styleDefEx(contentStyle, {
    marginBottom: localVariables.leftContentBottomMargin
});

// added by script - BEGIN
export const SidebarFooter = create;
// added by script - END
