// bobwai--viewer-container/src/styles.ts
import * as b from "bobril";

export const containerStyle = b.styleDef(
    {
        padding: "5px"
    },
    {},
    "viewer-container"
);

export const containerRelativePositionStyle = b.styleDef({
    position: "relative"
});

export const titleStyle = b.styleDef(
    {
        padding: "10px 0",
        fontSize: "17px",
        fontWeight: "bold"
    },
    {},
    "viewer-container-title"
);

export const subtitleStyle = b.styleDef(
    {
        padding: "5px 0",
        fontSize: "15px",
        fontWeight: "bold"
    },
    {},
    "viewer-container-subtitle"
);

export const descriptionStyle = b.styleDef(
    {
        padding: "5px 0"
    },
    {},
    "viewer-container-subtitle"
);

export const plainExampleStyle = b.styleDef(
    {
        border: "1px solid #ddd",
        background:
            "url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQAQMAAAAlPW0iAAAAAXNSR0IArs4c6QAAAAZQTFRF+vr6////+o8bnQAAABBJREFUCB1jYPjPgBXhEAYAfr8P8bm+jEoAAAAASUVORK5CYII=)"
    },
    {},
    "exampleFrame"
);

export const plainExampleDarkStyle = b.styleDefEx(plainExampleStyle, {
    background:
        "url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAoAAAAKCAYAAACNMs+9AAAAJ0lEQVQoU2MMCgpqYEAD7e3t6EIMjENB4c2bNzE8U1lZiemZIaAQAMD1JZD7MuRNAAAAAElFTkSuQmCC)"
});

export const paddingExampleStyle = b.styleDef(
    {
        paddingLeft: 10,
        paddingRight: 10,
        paddingTop: 10,
        paddingBottom: 10,
        borderRadius: "3px"
    },
    {},
    "exampleFrame"
);
