// bobwai--button/src/publicConstants.ts
import * as Color from "bobwai--color";
import * as ColorOperations from "bobwai--color-operations";

// note: this is intentionaly extracted, because OMS-500 has different "active elements" coloring (basic is used for hovered, application is not used)
let propTextThemeForegroundColor: Color.IColor = Color.Basic;
export function TextThemeForegroundColor(): string {
    return ColorOperations.makeFixedValue(propTextThemeForegroundColor);
}
export function setTextThemeForegroundColor(val: Color.IColor) {
    propTextThemeForegroundColor = val;
}

let propTextThemeForegroundDisabledColor: Color.IColor = Color.Basic_30;
export function TextThemeForegroundDisabledColor(): string {
    return ColorOperations.makeFixedValue(propTextThemeForegroundDisabledColor);
}
export function setTextThemeForegroundDisabledColor(val: Color.IColor) {
    propTextThemeForegroundDisabledColor = val;
}

let propTextThemeForegroundHoverColor: Color.IColor = Color.Application;
export function TextThemeForegroundHoverColor(): string {
    return ColorOperations.makeFixedValue(propTextThemeForegroundHoverColor);
}
export function setTextThemeForegroundHoverColor(val: Color.IColor) {
    propTextThemeForegroundHoverColor = val;
}

let textThemeColor: Color.IColor = Color.White;
export function TextThemeColor(): string {
    return ColorOperations.makeFixedValue(textThemeColor);
}
export function setTextThemeColor(val: Color.IColor) {
    textThemeColor = val;
}
