// bobwai--context-menu/examples/src/isTitleWithoutPaddingExample.ts
import * as b from "bobril";
import * as ContextMenu from "../../src/lib";
import * as Icon from "bobwai--icon";
import * as Color from "bobwai--color";

let menuItems: ContextMenu.IItem[] = [
    {
        id: "1",
        content: "View Template:",
        onClick: id => {
            alert("click " + id);
        },
        iconContent: Icon.create({ sprite: b.sprite(Icon.check_small_png, Color.Basic) })
    },
    {
        id: "2",
        content: "Clone",
        onClick: id => {
            alert("click " + id);
        },
        iconContent: Icon.create({ sprite: b.sprite(Icon.check_small_png, Color.Basic) })
    },
    {
        id: "3",
        content: "Edit",
        onClick: id => {
            alert("click " + id);
        },
        iconContent: Icon.create({ sprite: b.sprite(Icon.check_small_png, Color.Basic) })
    },
    {
        id: "4",
        content: "View Dependencies",
        isSelected: true,
        onClick: id => {
            alert("click " + id);
        },
        iconContent: Icon.create({ sprite: b.sprite(Icon.check_small_png, Color.Basic) })
    },
    {
        id: "5",
        content: "View Versions",
        onClick: id => {
            alert("click " + id);
        }
    },
    {
        content: "Submenu without icons",
        contextMenu: {
            title: "Title:",
            items: [
                {
                    content: "subItem 1"
                },
                {
                    content: "subItem 2"
                },
                {
                    content: "subItem 3"
                }
            ]
        }
    },
    {
        content: "Submenu with icons",
        contextMenu: {
            title: "Title:",
            items: [
                {
                    content: "subItem 1"
                },
                {
                    content: "subItem 2",
                    iconContent: Icon.create({ sprite: b.sprite(Icon.check_small_png, Color.Basic) })
                },
                {
                    content: "subItem 3"
                }
            ]
        }
    },
    {
        content: "Submenu with icons and title without padding",
        contextMenu: {
            title: "Title without padding:",
            items: [
                {
                    content: "subItem 1"
                },
                {
                    content: "subItem 2",
                    iconContent: Icon.create({ sprite: b.sprite(Icon.check_small_png, Color.Basic) })
                },
                {
                    content: "subItem 3"
                }
            ],
            isTitleWithoutPadding: true
        }
    }
];

let contextMenuNode: b.IBobrilChild = undefined;

let onRightClick = (event: b.IBobrilMouseEvent | undefined) => {
    contextMenuNode = ContextMenu.create({
        content: {
            title: "Tile without padding:",
            isTitleWithoutPadding: true,
            items: menuItems
        },
        mouseEvent: event,
        onClose: () => {
            contextMenuNode = undefined;
            b.invalidate();
        }
    });
    b.invalidate();
};

export default function() {
    return [
        contextMenuNode,
        b.style(
            {
                tag: "div",
                component: {
                    onMouseUp(_ctx: b.IBobrilCtx, event: b.IBobrilMouseEvent): boolean {
                        if (event.button === 3) {
                            onRightClick(event);
                            return true;
                        }
                        return false;
                    },
                    postInitDom(_ctx: b.IBobrilCtx, _me: b.IBobrilCacheNode, element: HTMLElement): void {
                        element.oncontextmenu = () => {
                            return false;
                        };
                    }
                }
            },
            { height: 200, background: "#0A99FF" }
        )
    ];
}
