// bobwai--l-view-app-header/src/lib.ts
import * as b from "bobril";
import * as styles from "./styles";
import * as HorizontalScrollCorrection from "bobwai--horizontal-scroll-correction";

export const create = b.createVirtualComponent<IData>({
    id: "bobwai--l-view-app-header",
    render(ctx: IContext, me: b.IBobrilNode): void {
        const headerContent = b.styledDiv(ctx.data.children, getHeaderStyle(ctx));

        me.children = [
            !ctx.data.isMobile &&
                HorizontalScrollCorrection.create({
                    content: headerContent,
                }),
            ctx.data.isMobile && headerContent,
        ];
    },
});

export interface IData {
    children?: b.IBobrilChildren;
    isMobile?: boolean;
    minWidth?: number;
}

export * from "./publicConstants";

interface IContext extends b.IBobrilCtx {
    data: IData;
}

function getHeaderStyle(ctx: IContext): b.IBobrilStyles {
    return [
        styles.baseStyle,
        ctx.data.isMobile && styles.baseMobileStyle,
        !ctx.data.isMobile && ctx.data.minWidth !== undefined && { minWidth: ctx.data.minWidth },
    ];
}

// added by script - BEGIN
export const LViewAppHeader = create;
// added by script - END
