// bobwai--example/src/colorPicker.ts
import * as b from "bobril";

export interface IData {
    availableValues: string[];
    onChange: (v: string) => void;
}

interface IContext extends b.IBobrilCtx {
    data: IData;
}

export const create = b.createComponent<IData>({
    id: "bobwai--example-color-picker",
    render(ctx: IContext, me: b.IBobrilNode) {
        me.children = ctx.data.availableValues.map((value) => {
            return b.style(
                {
                    tag: "div",
                    component: {
                        onClick(_c: b.IBobrilCtx, _e: b.IBobrilMouseEvent): boolean {
                            ctx.data.onChange(value);
                            return false;
                        },
                    },
                },
                exampleColorPickerItemStyle,
                { background: value }
            );
        });
        b.style(me, {});
    },
});

const exampleColorPickerItemStyle = b.styleDef({
    height: 20,
    width: 20,
    display: "inline-block",
    marginRight: 5,
    cursor: "pointer",
    borderRadius: 4,
});
