// bobwai--sidebar-item/src/action.ts
import * as b from "bobril";
import * as c from "./constants";
import * as d from "./data";
import * as s from "./styles";
import * as TooltipEmplacer from "bobwai--tooltip-emplacer";

export function create(data: d.IData): b.IBobrilChild {
    if (!data.actions) {
        return;
    }
    const children: b.IBobrilNode[] = [];
    if (d.isError(data) && !data.actionsVisibleOnlyOnHover) {
        children.push(
            TooltipEmplacer.create({
                content: b.styledDiv(c.errorIcon, actionItemStyle, actionItemFirstStyle),
                tooltipMessage: data.errorTooltip && data.errorTooltip.tooltipMessage,
                tooltipHeader: data.errorTooltip && data.errorTooltip.tooltipHeader,
                validationState: data.validationState,
            })
        );
    }
    data.actions.forEach((item: b.IBobrilChildren, index: number) => {
        children.push(b.styledDiv(item, actionItemStyle, index === 0 && actionItemFirstStyle));
    });
    return b.styledDiv(
        children,
        !data.actionsVisibleOnlyOnHover && actionStyle,
        data.actionsVisibleOnlyOnHover && [actionsParentHoverStyle, actionsParentNotHoverStyle]
    );
}

export const actionStyle = b.styleDef(
    {
        marginTop: (-1 * c.itemActionHeight) / 2,
        lineHeight: c.itemActionHeight + "px",
        display: "block",
        position: "absolute",
        right: c.itemActionRight,
        top: "50%",
        fontSize: 0,
    },
    {},
    "actionStyle"
);

const actionsParentHoverStyle = b.styleDefEx(s.baseStyle + ":hover>", actionStyle, {}, "actionsParentHoverStyle");

const actionsParentNotHoverStyle = b.styleDefEx(
    s.baseStyle + ":not(:hover)>",
    [actionStyle, { visibility: "hidden" }],
    {},
    "actionsParentNotHoverStyle"
);

const actionItemStyle = b.styleDef(
    {
        display: "inline-block",
        marginLeft: 9,
        fontSize: c.fontSize,
    },
    {},
    "actionItemStyle"
);

const actionItemFirstStyle = b.styleDefEx(
    actionItemStyle,
    {
        marginLeft: 7,
    },
    {},
    "actionItemFirstStyle"
);
