// bobwai--sidebar-item/src/constants.ts
import * as b from "bobril";
import * as Color from "bobwai--color";
import { Spacing } from "bobwai--form-elements";
import * as Icon from "bobwai--icon";

export const backgroundColor = Color.White;
export const invalidColor = Color.Error;
export const invalidActiveBackground = Color.Error_30;
export const warningColor = Color.Warning;
export const warningActiveBackground = Color.Warning_30;
export const paddingLeft = 25;
export const paddingRight = 25;
export const sidePaddingCJM = Spacing.Spacing500;
export const indentCircleWidth = 16;
export const indentWidth = 15 + indentCircleWidth;

export const borderColor = Color.Grey300;
export const hoverBackground = Color.Grey200;
export const hoverBorderColor = borderColor;
export const activeBackground: () => string = Color.Application_30;
export const activeBorderColor: () => string = Color.Application;
export const iconPlusMarginWidth = 30;
export const backgroundHover = 18;
export const counterColor: () => string = Color.Basic;
export const lineHeightNumber = 60;
export const lineHeight = lineHeightNumber + "px";
export const fontWeight = 700;
export const fontSize = 13;
export const color: () => string = Color.Basic;
export const border = "1px solid " + borderColor;

export const collapsibleIconHoverColor: () => string = Color.Basic;
export const collapsibleIconOpenColor: () => string = Color.Basic;

export const mediumLineHeight = "50px";
export const smallLineHeight = "45px";
export const smallMaxWidth = "95%";

export const separatorColor = "#515151"; // used if isSeparator: true (not in spec yet)

export const itemActionHeight = 20;
export const itemCollapsibleContentPadding = 25;
export const itemActionRight = 64;

export const errorIcon = Icon.create({ sprite: b.sprite(Icon.exclamation_small_png, Color.Error) });
