// bobwai--action-with-window/spec/GUI/layout.ts
import * as b from "bobril";
import { ActionWithWindow } from "../../src/lib";
import * as Color from "bobwai--color";
import * as Button from "bobwai--button";
import { getTestStore } from "./store";

export function createLayout(): void {
    b.init(() => {
        return [
            ActionWithWindow({
                actionContent: Button.create({
                    label: "label",
                    onClick: () => {},
                    propagateClick: true,
                }),
                windowContent: b.styledDiv("This is overlay window", {
                    paddingTop: 10,
                    paddingBottom: 10,
                    paddingLeft: 10,
                    paddingRight: 10,
                    width: 100,
                    height: 60,
                    background: Color.Basic(),
                }),
            }),
            ActionWithWindow({
                actionContent: Button.create({
                    label: "open",
                    onClick: () => {
                        getTestStore().isWindowVisible = !getTestStore().isWindowVisible;
                        b.invalidate();
                    },
                    propagateClick: true,
                }),
                windowContent: b.styledDiv(
                    Button.create({
                        label: "close",
                        onClick: () => {
                            getTestStore().isWindowVisible = !getTestStore().isWindowVisible;
                            b.invalidate();
                        },
                    }),
                    {
                        paddingTop: 10,
                        paddingBottom: 10,
                        paddingLeft: 10,
                        paddingRight: 10,
                        width: 100,
                        height: 60,
                        background: Color.Basic(),
                    }
                ),
                isWindowVisible: getTestStore().isWindowVisible,
            }),
        ];
    });
}
