// bobwai--color/examples/src/Container.ts
import * as b from "bobril";

interface IContext extends b.IBobrilCtx {
    data: IData;
}

export interface IData {
    header: string;
    content: b.IBobrilChildren;
}

export let create = b.createComponent<IData>({
    render(ctx: IContext, me: b.IBobrilNode): void {
        //b.style(me, styles.containerStyle);
        me.children = [b.styledDiv(ctx.data.header, titleStyle), b.styledDiv(ctx.data.content, plainExampleStyle)];
    }
});

export const titleStyle = b.styleDef(
    {
        padding: "10px 0",
        fontSize: "17px",
        fontWeight: "bold"
    },
    {},
    "color-container"
);

export const containerStyle = b.styleDef(
    {
        padding: "5px"
    },
    {},
    "color-container"
);

export const plainExampleStyle = b.styleDef(
    {
        border: "1px solid #ddd",
        background:
            "url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQAQMAAAAlPW0iAAAAAXNSR0IArs4c6QAAAAZQTFRF+vr6////+o8bnQAAABBJREFUCB1jYPjPgBXhEAYAfr8P8bm+jEoAAAAASUVORK5CYII=)"
    },
    {},
    "exampleFrame"
);
