// bobwai--form-elements/src/lib.ts
import { ContentTopIndent } from "./contentTopIndentStyles";
export { ContentTopIndent, getContentTopIndentStyle } from "./contentTopIndentStyles";
export { OnChangeDefaultDelay } from "./onChangeDelay";
export { Spacing } from "bobwai--style-mixins";

export enum ValidationState {
    Default,
    Error,
    Warning,
}

export enum LabelWidth {
    Label100 = 50,
    Label150 = 80,
    Label200 = 100,
    Label300 = 120,
    Label350 = 140,
    Label400 = 160,
    Label500 = 180,
    Label600 = 220,
    Label700 = 280,
}

export interface ITooltipData {
    tooltipMessage?: string | string[];
    tooltipHeader?: string | string[];
    /**
     * Using for case when some of parent elements contain scroll and can cover part of content.
     */
    computePositionFromVisibleContent?: boolean;
}

export interface IBorderVisibility {
    hasBorderTop?: boolean;
    hasBorderBottom?: boolean;
    hasBorderLeft?: boolean;
    hasBorderRight?: boolean;
}

export enum DisplayAs {
    Default,
    Block,
    InlineBlock,
    Table,
}

export enum VerticalAlign {
    Default,
    Top,
    Middle,
}

export enum TextAlign {
    Left,
    Center,
    Right,
}

export enum SiblingType {
    None,
    Left,
    Right,
    Both,
}

export interface IContentSettings {
    variant?: ContentVariant;
    topIndent?: ContentTopIndent;
    isFullSize?: boolean;
}

export enum ContentVariant {
    Default,
    FullWidth,
}

export type ColumnCount = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12;
