// bobwai--scroll-shadow/src/styles.ts
import * as b from "bobril";
import * as constants from "./publicConstants";
import * as Color from "bobwai--color";

// general
const shadowHeight = 10;
export const shadowBoundary = 42;
export const sidePadding = constants.SidePadding;
const sidePaddingForAsterisks = 10;
const bottomBorderHeight = 1;

// images
const topShadowImage =
    "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAAKCAYAAAB10jRKAAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAAJcE" +
    "hZcwAADsMAAA7DAcdvqGQAAAAYdEVYdFNvZnR3YXJlAHBhaW50Lm5ldCA0LjAuOWwzfk4AAAAmSURBVBhXY5g8eXIliIgFES4gQh9EyIMIYRDBBSJYQQQTkJjMCADxBRLIi" +
    "6jhSQAAAABJRU5ErkJggg==";
const bottomShadowImage =
    "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAAKCAYAAAB10jRKAAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAA" +
    "JcEhZcwAADsMAAA7DAcdvqGQAAAAYdEVYdFNvZnR3YXJlAHBhaW50Lm5ldCA0LjAuOWwzfk4AAAAmSURBVBhXY5g8eTIjiGACEawgggtECIMIeRChDyJcQEQskJhcCQDG7h" +
    "LISyKEFAAAAABJRU5ErkJggg==";

const topCornerLeftImage =
    "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACoAAAAKCAYAAADCb8xtAAAABGdBTUEAALGPC/xhBQAAAAlwSFlzAAAOwgAADs" +
    "IBFShKgAAAABh0RVh0U29mdHdhcmUAcGFpbnQubmV0IDQuMC45bDN+TgAAAX5JREFUOE+90sdKBFEURVHbnHPOWTG1GQOKWTCgmDA1KDgQFEVExO+oH3Yfebd5FNXVDsTBo" +
    "qZn31c5QRBUohaNaEEHujGAIYxgArOYxyKWsYp1bGMHu9jDAY5xijNc4BLXSOEO93hwHvGEZ7zgFW94xwc+8aWhpShHtVOPJrSiDV3oxyCGoeGjSGIaM5iDIhawBAtZg2I2" +
    "oCCL2cehc4QTKOocFnaFG9wipaEFKEQxyqDRunINdGmN1qU1vB2d6HH6oMsrQtcXxShkDHoJmcQUFGYvoygLW3EsTmGy6WxpaJ6TD40ugo2ugEbr0hotungD9KsootmxGL2" +
    "Cfh8F6TUsqhcWZnH+K9lLKXDcUaS+SQ3NRXisf2UpgX4RG6+rW0QV7LcRe4m6ED8uHOhH6tVMOlhDE44GGxvuj9dw0cV94ZgwxRkL9CONH2rSwf7QsKjR/vhM7EXCLDRbrE" +
    "9xP8FxQ+P41/+tqGCTNVRDo0SN+2tRMRlFjfwPUcNjBIlvWr/6Jt/songAAAAASUVORK5CYII=";

const topCornerRightImage =
    "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACoAAAAKCAYAAADCb8xtAAAABGdBTUEAALGPC/xhBQAAAAlwSFlzAAAOwgAAD" +
    "sIBFShKgAAAABh0RVh0U29mdHdhcmUAcGFpbnQubmV0IDQuMC45bDN+TgAAAY1JREFUOE+t0VdLA1EUReFYYokae+9d7B0Lil2woNiwgoIPgqKIiPg78oPj2uGecBnHxIc8" +
    "fPhiZq85E0kkEt/4wic+8I43vOIFz3jCo/OAe9zhGpe4wBlOcYwD7GEXO9jGOlaxjEXMYxbjGMYg+tGFdjSjATWIK1SDt7jBFWz4HBo/wREOnX1YhAI2oIg1WMgSFqCYOcx" +
    "gGhMYgcKGMIA+dKIVLWhEHaqccsQUuoVNR6NioyuODdu4LqHhKUxCV5FRKEQRupAoRpfqRbfTgTYoTJdTnC5XjTgUV4YSFCGqUL3lGDSkv2KDenP/7W3Uhntg47qKAvTZ7D" +
    "qKaHIUo09ZD11MYaKrKa4CFleMKApRIAr1B/SWxh8LDvqjtQF2Gft0UgmL0bUsKoZSKC59PScdiXyFBgeMP2JszAY1ZjQa5EfoSj5FWZgflQrz5IlC7aEa04N92Ub8CwRp/" +
    "C9+VDAuFRak0LARyfZg47/9f4XGZKLQsAdlEvqgXEomk5EghYb+cxb6XU6ERf2WjPwAbH8NMYAcQVoAAAAASUVORK5CYII=";

const bottomCornerLeftImage =
    "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACoAAAAKCAYAAADCb8xtAAAABGdBTUEAALGPC/xhBQAAAAlwSFlzAAAOwgA" +
    "ADsIBFShKgAAAABh0RVh0U29mdHdhcmUAcGFpbnQubmV0IDQuMC45bDN+TgAAAXBJREFUOE/F0ddKJGEURWFz1jGMjjknHHPCgGIeMKCYMIKCF4Iyg4iIz1EvPGvLOc1PUb" +
    "a2CF589E1Ta9eprCiKvkN2ppIeIol//mI5mfjsqMSHvSM3jbxAfpJ0Q9MFwgfHJYZQECiMKTLFMaUokXDoW2/v8Y9EXh8ao5grQ7mpwI9ApakKVKNGNDQcFl5Eo+IjFFPEg" +
    "x7ziIdSgUAt6vDL1JsG04gmNAda0Io2DY2P9Cv5KI3RAMXlJ8JoGPRYKoAO04ku9Jhe048B8xuDGDLD9juioeH1/PNonF9GY3yE3lIDPO5hBfuMwh5USEYxhnFMYhozZg4L" +
    "ZhFLWDGrZk1D9Uk1zj+dLqZxfh1dpRvh22vICBSewBQ8rvA8PLoMRdfxB5vYwo7ZxT4OcYRjnOAU57jApYbqerqcPqUup8/WDl1KF9IwXcUvMQsfohEasAEfsY09HEBxD5/" +
    "hEle4xo25xR3u8Rf/8IBHPOE5iqKX/zb/+ib5NebGAAAAAElFTkSuQmCC";

const bottomCornerRightImage =
    "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACoAAAAKCAYAAADCb8xtAAAABGdBTUEAALGPC/xhBQAAAAlwSFlzAAAOwg" +
    "AADsIBFShKgAAAABh0RVh0U29mdHdhcmUAcGFpbnQubmV0IDQuMC45bDN+TgAAAXZJREFUOE+t0lVLRGEURmG7u7sLuzFQbMFAsTBBwQtBUURE/B3zg8f1wt7yOZwZAy8e5" +
    "kbhXd8+abFYLP0P9H//Ih6P/4j+OOOXoob/q2RDs5PICmSmEBXznciBqWhogSlEfoI8k5sgJxAVKWFoolTBSYdWmgqUB8pMaaAExaYIinMeHPpJrEdFjf4crqGtaEEzmgKN" +
    "aEC9qTO1pgbV8NDEYA/1WI9UoH49LowJLxSOz9DQEQxh2H5lEAPoN33oMd2mEx1oN2FwGBkGelwVFCQK8QiN99f/MlZD17BqVswSFrFg5jBjpjGJcYxhFIoUD1RYr/E4hXm" +
    "UgnQ1j1GEX0Kj/bP6fGUNvcIlLnCGU5zgGEc4wB52zTa2sIl1KGwZHjcPD1PUFCagMF3PL+VX6oKu4VfQaL24fzYaXaCh73jDK17wjCc84gH3uMOtucE1FHgOD1PUIfaxA4" +
    "/ZgIIU4yGz8MvoEhqu19fLt0Gfj15anwovHSv5AL8rDTGdNw4EAAAAAElFTkSuQmCC";

// styles
export const fullHeightStyle = b.styleDef(
    {
        height: "100%",
    },
    {},
    "scroll-full-height"
);

export const wrapperStyle = b.styleDef(
    {
        position: "relative",
    },
    {},
    "scroll-wrapper"
);

export const overflowHiddenStyle = b.styleDef({
    overflow: "hidden",
});

export const contentStyle = b.styleDef(
    {
        overflow: "auto",
    },
    {},
    "scroll-content"
);

export const contentRelativeStyle = b.styleDef(
    {
        position: "relative",
    },
    {},
    "scroll-content-relative"
);

export const contentSidePaddingStyle = b.styleDef(
    {
        paddingLeft: sidePadding,
        paddingRight: sidePadding,
    },
    {},
    "scroll-content-with-padding"
);

export const contentSidePaddingForAsterisksStyle = b.styleDef(
    {
        marginLeft: -1 * sidePaddingForAsterisks,
        paddingLeft: sidePaddingForAsterisks,
    },
    {},
    "scroll-content-with-padding-for-asterisks"
);

export const shadowStyle = b.styleDef(
    {
        position: "absolute",
        opacity: 0,
        height: shadowHeight,
        marginLeft: shadowBoundary,
        marginRight: shadowBoundary,
        backgroundRepeat: "repeat-x",
        pointerEvents: "none",
    },
    {
        before: [
            {
                position: "absolute",
                display: "block",
                content: "",
                top: 0,
                left: -shadowBoundary,
                width: shadowBoundary,
                height: shadowHeight,
                pointerEvents: "none",
            },
        ],
        after: [
            {
                position: "absolute",
                display: "block",
                content: "",
                top: 0,
                right: -shadowBoundary,
                width: shadowBoundary,
                height: shadowHeight,
                pointerEvents: "none",
            },
        ],
    },
    "scroll-shadow"
);

export const topShadowStyle = b.styleDef(
    {
        top: 0,
        backgroundImage: "url(" + topShadowImage + ")",
    },
    {
        before: {
            backgroundImage: "url(" + topCornerLeftImage + ")",
        },
        after: {
            backgroundImage: "url(" + topCornerRightImage + ")",
        },
    },
    "scroll-shadow-top"
);

export const bottomShadowStyle = b.styleDef(
    {
        bottom: 0,
        backgroundImage: "url(" + bottomShadowImage + ")",
    },
    {
        before: {
            backgroundImage: "url(" + bottomCornerLeftImage + ")",
        },
        after: {
            backgroundImage: "url(" + bottomCornerRightImage + ")",
        },
    },
    "scroll-shadow-bottom"
);

// just for identification in DOM
export const scrollShadowWithOffsetWrapperStyle = b.styleDef({});

export const scrollShadowWithOffsetWrapperFullSizeStyle = b.styleDef({
    height: "100%",
});

export const scrollShadowOffsetIECorrection = b.styleDef({ paddingBottom: 1, marginBottom: -1 });

export const shadowLeftSidePaddingStyle = b.styleDefEx(shadowStyle, {
    left: sidePadding,
});

export const scrollShadowBottomBorderStyle = b.styleDef(
    {
        position: "absolute",
        bottom: -bottomBorderHeight,
        left: 0,
        right: 0,
        height: 1,
        background: Color.Grey300,
    },
    {},
    "scroll-shadow-bottom-border"
);

export const scrollShadowWithBottomBorderStyle = b.styleDef(
    {
        paddingBottom: bottomBorderHeight,
        boxSizing: "border-box",
    },
    {},
    "scroll-shadow-with-bottom-border"
);
