// bobwai--scroll-shadow/src/styles.ts
import * as b from "bobril";
import * as constants from "./publicConstants";
import * as Color from "bobwai--color";

// general
const shadowHeight = 10;
export const shadowBoundary = 42;
export const sidePadding = constants.SidePadding;
const sidePaddingForAsterisks = 10;
const bottomBorderHeight = 1;

// images
const topShadowImage =
    "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAAKCAYAAAB10jRKAAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAAJcE" +
    "hZcwAADsMAAA7DAcdvqGQAAAAYdEVYdFNvZnR3YXJlAHBhaW50Lm5ldCA0LjAuOWwzfk4AAAAmSURBVBhXY5g8eXIliIgFES4gQh9EyIMIYRDBBSJYQQQTkJjMCADxBRLIi" +
    "6jhSQAAAABJRU5ErkJggg==";

const bottomShadowImage =
    "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAAKCAYAAAB10jRKAAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAA" +
    "JcEhZcwAADsMAAA7DAcdvqGQAAAAYdEVYdFNvZnR3YXJlAHBhaW50Lm5ldCA0LjAuOWwzfk4AAAAmSURBVBhXY5g8eTIjiGACEawgggtECIMIeRChDyJcQEQskJhcCQDG7h" +
    "LISyKEFAAAAABJRU5ErkJggg==";

const topCornerLeftImage =
    "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACoAAAAKCAYAAADCb8xtAAAABGdBTUEAALGPC/xhBQAAAAlwSFlzAAAOwgAADs" +
    "IBFShKgAAAABh0RVh0U29mdHdhcmUAcGFpbnQubmV0IDQuMC45bDN+TgAAAX5JREFUOE+90sdKBFEURVHbnHPOWTG1GQOKWTCgmDA1KDgQFEVExO+oH3Yfebd5FNXVDsTBo" +
    "qZn31c5QRBUohaNaEEHujGAIYxgArOYxyKWsYp1bGMHu9jDAY5xijNc4BLXSOEO93hwHvGEZ7zgFW94xwc+8aWhpShHtVOPJrSiDV3oxyCGoeGjSGIaM5iDIhawBAtZg2I2" +
    "oCCL2cehc4QTKOocFnaFG9wipaEFKEQxyqDRunINdGmN1qU1vB2d6HH6oMsrQtcXxShkDHoJmcQUFGYvoygLW3EsTmGy6WxpaJ6TD40ugo2ugEbr0hotungD9KsootmxGL2" +
    "Cfh8F6TUsqhcWZnH+K9lLKXDcUaS+SQ3NRXisf2UpgX4RG6+rW0QV7LcRe4m6ED8uHOhH6tVMOlhDE44GGxvuj9dw0cV94ZgwxRkL9CONH2rSwf7QsKjR/vhM7EXCLDRbrE" +
    "9xP8FxQ+P41/+tqGCTNVRDo0SN+2tRMRlFjfwPUcNjBIlvWr/6Jt/songAAAAASUVORK5CYII=";

const topCornerRightImage =
    "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACoAAAAKCAYAAADCb8xtAAAABGdBTUEAALGPC/xhBQAAAAlwSFlzAAAOwgAAD" +
    "sIBFShKgAAAABh0RVh0U29mdHdhcmUAcGFpbnQubmV0IDQuMC45bDN+TgAAAY1JREFUOE+t0VdLA1EUReFYYokae+9d7B0Lil2woNiwgoIPgqKIiPg78oPj2uGecBnHxIc8" +
    "fPhiZq85E0kkEt/4wic+8I43vOIFz3jCo/OAe9zhGpe4wBlOcYwD7GEXO9jGOlaxjEXMYxbjGMYg+tGFdjSjATWIK1SDt7jBFWz4HBo/wREOnX1YhAI2oIg1WMgSFqCYOcx" +
    "gGhMYgcKGMIA+dKIVLWhEHaqccsQUuoVNR6NioyuODdu4LqHhKUxCV5FRKEQRupAoRpfqRbfTgTYoTJdTnC5XjTgUV4YSFCGqUL3lGDSkv2KDenP/7W3Uhntg47qKAvTZ7D" +
    "qKaHIUo09ZD11MYaKrKa4CFleMKApRIAr1B/SWxh8LDvqjtQF2Gft0UgmL0bUsKoZSKC59PScdiXyFBgeMP2JszAY1ZjQa5EfoSj5FWZgflQrz5IlC7aEa04N92Ub8CwRp/" +
    "C9+VDAuFRak0LARyfZg47/9f4XGZKLQsAdlEvqgXEomk5EghYb+cxb6XU6ERf2WjPwAbH8NMYAcQVoAAAAASUVORK5CYII=";

const bottomCornerLeftImage =
    "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACoAAAAKCAYAAADCb8xtAAAABGdBTUEAALGPC/xhBQAAAAlwSFlzAAAOwgA" +
    "ADsIBFShKgAAAABh0RVh0U29mdHdhcmUAcGFpbnQubmV0IDQuMC45bDN+TgAAAXBJREFUOE/F0ddKJGEURWFz1jGMjjknHHPCgGIeMKCYMIKCF4Iyg4iIz1EvPGvLOc1PUb" +
    "a2CF589E1Ta9eprCiKvkN2ppIeIol//mI5mfjsqMSHvSM3jbxAfpJ0Q9MFwgfHJYZQECiMKTLFMaUokXDoW2/v8Y9EXh8ao5grQ7mpwI9ApakKVKNGNDQcFl5Eo+IjFFPEg" +
    "x7ziIdSgUAt6vDL1JsG04gmNAda0Io2DY2P9Cv5KI3RAMXlJ8JoGPRYKoAO04ku9Jhe048B8xuDGDLD9juioeH1/PNonF9GY3yE3lIDPO5hBfuMwh5USEYxhnFMYhozZg4L" +
    "ZhFLWDGrZk1D9Uk1zj+dLqZxfh1dpRvh22vICBSewBQ8rvA8PLoMRdfxB5vYwo7ZxT4OcYRjnOAU57jApYbqerqcPqUup8/WDl1KF9IwXcUvMQsfohEasAEfsY09HEBxD5/" +
    "hEle4xo25xR3u8Rf/8IBHPOE5iqKX/zb/+ib5NebGAAAAAElFTkSuQmCC";

const bottomCornerRightImage =
    "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACoAAAAKCAYAAADCb8xtAAAABGdBTUEAALGPC/xhBQAAAAlwSFlzAAAOwg" +
    "AADsIBFShKgAAAABh0RVh0U29mdHdhcmUAcGFpbnQubmV0IDQuMC45bDN+TgAAAXZJREFUOE+t0lVLRGEURmG7u7sLuzFQbMFAsTBBwQtBUURE/B3zg8f1wt7yOZwZAy8e5" +
    "kbhXd8+abFYLP0P9H//Ih6P/4j+OOOXoob/q2RDs5PICmSmEBXznciBqWhogSlEfoI8k5sgJxAVKWFoolTBSYdWmgqUB8pMaaAExaYIinMeHPpJrEdFjf4crqGtaEEzmgKN" +
    "aEC9qTO1pgbV8NDEYA/1WI9UoH49LowJLxSOz9DQEQxh2H5lEAPoN33oMd2mEx1oN2FwGBkGelwVFCQK8QiN99f/MlZD17BqVswSFrFg5jBjpjGJcYxhFIoUD1RYr/E4hXm" +
    "UgnQ1j1GEX0Kj/bP6fGUNvcIlLnCGU5zgGEc4wB52zTa2sIl1KGwZHjcPD1PUFCagMF3PL+VX6oKu4VfQaL24fzYaXaCh73jDK17wjCc84gH3uMOtucE1FHgOD1PUIfaxA4" +
    "/ZgIIU4yGz8MvoEhqu19fLt0Gfj15anwovHSv5AL8rDTGdNw4EAAAAAElFTkSuQmCC";

const leftShadowImage =
    "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAoAAAABCAYAAADn9T9+AAACyHpUWHRSYXcgcHJvZmlsZSB0eXBlIGV4aWYAAHjaxZVtjhwpDIb/c4ocAdsYm" +
    "+NQfEh7gz1+Xiime3rSiZRkpcVTQLldBt4HmDD+/WeGbygUhUNS81xyjiippMIVHY93uVuKadd3qaelV394BDB6glbu1zxOfIVfnx9YOv7r1R+snTx+En0kPglljczonDg/" +
    "iYRvP533UD5mnD4t5zxt7BSRTtKv78kgRlc4oREPgR91XqPI/dTl23VBEMG3+gl1lPReu/DofhGPynvtYj0R8ipFiPkE5C8aHT/pe+22Qp9nRB9dfv1h8q3Ho3zSbs7uc45" +
    "7dTVlKJXDWVQ8KXYPgReklP1ZhhkeRd+2FZhjiQ3EOmhesBaoEEPZSYk6VZo0dtuoYUqJBxta5sayfS7GhZvELT6MJpsU6UEcTBqoCdz8mAvtccser5Fj5E6IZEKyRfEHC+" +
    "+cf2KPRHOurUu0xAR6ugHzEhzTWORWjSgAoXk01a3vtudZe5YFVkBQt8yOBdZ43Skupefeks1ZEKcxhXgfDbJ+EkAijK2YDAkIxEyilCkasxFBRwefipmzJL5AgFS5U5hgI" +
    "5IBx3mNjW+Mdiwr325cLQChksWApkgFrJQU+8eSYw9VFU1BVbOauhatWXLKmnO2vO6oamLJ1LKZuRWrLp5cPbu5e/FauAiuMC25WCheSqkVg1akrvi6IqLWiy+50qVXvuzy" +
    "q1y1Yfu01LTlZs1babVzl47j33O30L2XXgcNbKWRho48bPgoo07stSkzTZ152vRZZn1QO1RfqdEXcr+mRofaIpZ2nD2pwW32kYLWdaKLGYhxIhC3RQAbmhez6JQSL3KLWSy" +
    "MQ6EMaqQLTqdFDATTINZJD3ZPcr/kFjT9Fjf+Gbmw0P0X5MJCd8j9yO0NtV73fxTZgNYpXJpGmbjYulR2/OE6/os2/G2C/zWRQKESvgNNn9gp/5Y97AAAAAZiS0dEAP8A/w" +
    "D/oL2nkwAAAAlwSFlzAAAOwwAADsMBx2+oZAAAAAd0SU1FB+QKEwsOB0WvRCgAAAAoSURBVAjXBcExDgAQAASwerZ33BclBotBYhChLUkqGjoGJhY2Di7eB+KyCrxBRf2CA" +
    "AAAAElFTkSuQmCC";

const rightShadowImage =
    "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAoAAAABCAYAAADn9T9+AAACx3pUWHRSYXcgcHJvZmlsZSB0eXBlIGV4aWYAAHjaxZVtjhwpDIb/c4ocAdsYm" +
    "+NQfEh7gz1+Xiime3rSiZRkpcVTQLldBt4HmDD+/WeGbygUhUNS81xyjiippMIVHY93uVuKadd3qaelV394BDB6glbu1zxOfIVfnx9YOv7r1R+snTx+En0kPglljczonDg/" +
    "iYRvP533UD5mnD4t5zxt7BSRTtKv78kgRlc4oREPgR91XqPI/dTl23VBEMG3+gm1CL3XLjy6X8Sj8l67WE+EvEoRYj4B+YtGx0/6Xrut0OcZ0UeXX3+YfOvxKJ+0m7P7nON" +
    "eXU0ZSuVwFhVPit1D4AUpZX+WYYZH0bdtBeZYYgOxDpoXrAUqxFB2UqJOlSaN3TZqmFLiwYaWubFsn4tx4SZxiw+jySZFehAHkwZqAjc/5kJ73LLHa+QYuRMimZBsUfzBwj" +
    "vnn9gj0Zxr6xItMYGebsC8BMc0FrlVIwpAaB5Ndeu77XnWnmWBFRDULbNjgTVed4pL6bm3ZHMWxGlMId5Hg6yfBJAIYysmQwICMZMoZYrGbETQ0cGnYuYsiS8QIFXuFCbYi" +
    "GTAcV5j4xujHcvKtxtXC0CoZDGgKVIBKyXF/rHk2ENVRVNQ1aymrkVrlpyy5pwtrzuqmlgytWxmbsWqiydXz27uXrwWLoIrTEsuFoqXUmrFoBWpK76uiKj14kuudOmVL7v8" +
    "Kldt2D4tNW25WfNWWu3cpeP499wtdO+l10EDW2mkoSMPGz7KqBN7bcpMU2eeNn2WWR/UDtVXavSF3K+p0aG2iKUdZ09qcJt9pKB1nehiBmKcCMRtEcCG5sUsOqXEi9xiFgv" +
    "jUCiDGumC02kRA8E0iHXSg92T3C+5BU2/xY1/Ri4sdP8FubDQHXI/cntDrdf9H0U2oHUKl6ZRJi62LpUdf7iO/6INf5vgf00kUKiE702l2CmGsAl2AAAABmJLR0QA/wD/AP" +
    "+gvaeTAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAAB3RJTUUH5AoTCw4hl6LB1QAAACZJREFUCNcFwTEKACAMBLAc4iA4uHQQ/HU/bJLuDoKBiYWNg8LF+0hKAjQfrNgwAAAAA" +
    "ElFTkSuQmCC";

const leftCornerTopImage =
    "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAoAAAAqCAYAAACKnsqLAAACznpUWHRSYXcgcHJvZmlsZSB0eXBlIGV4aWYAAHja7ZVtctwgDIb/c4oewZIQE" +
    "sfBfMz0Bj1+XzC7m022mWnTmf6piQFrhST0CBL6j+8jfMNDlI4Q1TzllA48McfMBRM/ruca6Yirv56yR3qWB779wBAJRrk+U9/6BXJ9LLC45eezPFjddnwbunneBmV6Zky2" +
    "nm9Dwpec9nfIt4jjm+3sN9Fhe9EV5bvvaEhGU9gTDtyF5ECfpoJcb5my1WcoEWRznlbPr3MXpL5O3n32LndH2XJ5TkU40lZI73K05aSvc7cy9DYienh++sH4vuRD7sZoPka" +
    "/dldiQqZS2Ju6bWXNoHgilbKWJTTDq5jbahnNscUKYg00T7QaKBMjs4MiNSo0qK+xUkWIkTsjNGauLEvmYpy5ykQQZ6PBJllaEAeTCmoCMd9joeU3L3+VHJ4bQZMJxibFDy" +
    "28Ev5JuxsaY9YB0eFXnlAWiItn1SCMSW720AIQGjunuvK7WnhTN8cbsAKCutLs2GA5zsvEqfSoLVmcBXp6xHBcR4OsbQNIEXwrgiEBAZwLUZqHg9mIkEcHn4LIWSKfIECq3" +
    "CgMsBFJgOM8fWON0dJl5UuMqwUgFEfEgCZLAawYFfVj0VFDRUVjUNWkpq5ZS5IUk6aULM07qphYNLVkZm7ZiotHV09u7p69ZM6CK0xzyhay55xLgdMC0wWrCzRKOfmUM556" +
    "ptNOP/NZKsqnxqo1Vatecy2NmzQc/5aaheYtt9Kpo5R67NpTt+499zJQa0NGHDrSsOEjj3KnRvvYPlGjd+Q+p0ab2iQWl549qEFsdjNB8zrRyQzEOBKI2ySAgubJ7HCKkSe" +
    "5yezIjEOhDGqkE06jSQwEYyfWQXd2D3Kfcgsaf4sb/4pcmOj+Brkw0W1yH7m9oNbK+o8iC9A8hTOnhwxcbE0KO/5wHX9hDF818N/QPzYkqJIcfgLjXRbJNyKj1AAAAAZiS0" +
    "dEAP8A/wD/oL2nkwAAAAlwSFlzAAAOwgAADsIBFShKgAAAAAd0SU1FB+QKEwsQDXE7kukAAADfSURBVDjLjZTbDoMwDENtl+1xP8/nblr3UqQs5NJKCBDmNImT8jzPF4APg" +
    "DeAL4C5rmtNABCAY925PhDBuoStWADGeh5L0BLpqCFxFFvzEj6MeGRxypBo3tOtj0W2NEbJPAL6jfg0glIoF1tZHv9DSqQpzYwSslmjs9A7U3rtMw5j5G4/MiCxKg86oZL4" +
    "mCUzdhuXXXlYWRdZiKox5KjoiMONQ9oUWeHTplDndRTj32wroYUHQLSY9aO6ossdJ3JURhaimhvtNq8vTyrWTiLYEXSjcHPKE6cRzu7EDcP4AcutDHVizXNYAAAAAElFTkS" +
    "uQmCC";

const leftCornerBottomImage =
    "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAoAAAAqCAYAAACKnsqLAAACznpUWHRSYXcgcHJvZmlsZSB0eXBlIGV4aWYAAHja7ZVtctwgDIb/c4oewZIQE" +
    "sfBfMz0Bj1+XzC7m022mWnTmf6piQFrhST0CBL6j+8jfMNDlI4Q1TzllA48McfMBRM/ruca6Yirv56yR3qWB779wBAJRrk+U9/6BXJ9LLC45eezPFjddnwbunneBmV6Zky2" +
    "nm9Dwpec9nfIt4jjm+3sN9Fhe9EV5bvvaEhGU9gTDtyF5ECfpoJcb5my1WcoEWRzruij5Ne5C1JfJ+8+e5e7o2y5PKciHGkrpHc52nLS17lbGXobET08P/1gfF/yIXdjNB+" +
    "jX7srMSFTKexN3bayZlA8kUpZyxKa4VXMbbWM5thiBbEGmidaDZSJkdlBkRoVGtTXWKkixMidERozV5YlczHOXGUiiLPRYJMsLYiDSQU1gZjvsdDym5e/Sg7PjaDJBGOT4o" +
    "cWXgn/pN0NjTHrgOjwK08oC8TFs/QQxiQ3e2gBCI2dU135XS28qZvjDVgBQV1pdmywHOdl4lR61JYszgI9PWI4rqNB1rYBpAi+FcGQgADOhSjNw8FsRMijg09B5CyRTxAgV" +
    "W4UBtiIJMBxnr6xxmjpsvIlxtUCECpJDGiyFMCKUVE/Fh01VFQ0BlVNauqatSRJMWlKydK8o4qJRVNLZuaWrbh4dPXk5u7ZS+YsuMI0p2whe865FDgtMF2wukCjlJNPOeOp" +
    "Zzrt9DOfpaJ8aqxaU7XqNdfSuEnD8W+pWWjeciudOkqpx649devecy8DtTZkxKEjDRs+8ih3arSP7RM1ekfuc2q0qU1icenZgxrEZjcTNK8TncxAjCOBuE0CKGiezA6nGHm" +
    "Sm8yOzDgUyqBGOuE0msRAMHZiHXRn9yD3Kbeg8be48a/IhYnub5ALE90m95HbC2qtrP8osgDNUzhzesjAxdaksOMP1/EXxvBVA/8N/WNDgirJ4ScqKBbSZPLXfAAAAAZiS0" +
    "dEAP8A/wD/oL2nkwAAAAlwSFlzAAAOwgAADsIBFShKgAAAAAd0SU1FB+QKEwsPM30AgdwAAAD1SURBVDjLjZTBigQhDERf+9f7Hf2LA3PYyx4WBmZMshddJG3SEQTBSkxZl" +
    "RzneX4BD+AJfAM/wC/wAj6AANYABWzseb6s5i6PJejYAdesa/YwY7jWjGlA2xCxJcB2T4egNeNauO1KaOND5w4DGtArhFpAMnxaA9b/QS1SwoObI6J3rKO/vEgomxIsUkYr" +
    "ppiRmn2PVd0jrsZQmT4updIzWrGZbEhY5nCrADUDeTJ6J6Fs9A7/sWf94iUUN0ouphDXM5rZTDNDTODbWS1k3cfALLXCBFvEfjdxNarx48ZK6h5J9LaVTK8MexlnqbRCv7N" +
    "aq4AA/gBo09emaGyWlwAAAABJRU5ErkJggg==";

const rightCornerTopImage =
    "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAoAAAAqCAYAAACKnsqLAAACz3pUWHRSYXcgcHJvZmlsZSB0eXBlIGV4aWYAAHja7ZVtctwgDIb/c4oewZIQE" +
    "sfBfMz0Bj1+XzC7m022mWnTmf6piQFrhST0CBL6j+8jfMNDlI4Q1TzllA48McfMBRM/ruca6Yirv56yR3qWB779wBAJRrk+U9/6BXJ9LLC45eezPFjddnwbunneBmV6Zky2" +
    "nm9Dwpec9nfIt4jjm+3sN9Fhe9EV5bvvaEhGU9gTDtyF5ECfpoJcb5my1WcoEWRzntCL5Ne5C1JfJ+8+e5e7o2y5PKciHGkrpHc52nLS17lbGXobET08P/1gfF/yIXdjNB+" +
    "jX7srMSFTKexN3bayZlA8kUpZyxKa4VXMbbWM5thiBbEGmidaDZSJkdlBkRoVGtTXWKkixMidERozV5YlczHOXGUiiLPRYJMsLYiDSQU1gZjvsdDym5e/Sg7PjaDJBGOT4o" +
    "cWXgn/pN0NjTHrgOjwK08oC8TFs/QQxiQ3e2gBCI2dU135XS28qZvjDVgBQV1pdmywHOdl4lR61JYszgI9PWI4rqNB1rYBpAi+FcGQgADOhSjNw8FsRMijg09B5CyRTxAgV" +
    "W4UBtiIJMBxnr6xxmjpsvIlxtUCEIojYkCTpQBWjIr6seiooaKiMahqUlPXrCVJiklTSpbmHVVMLJpaMjO3bMXFo6snN3fPXjJnwRWmOWUL2XPOpcBpgemC1QUapZx8yhlP" +
    "PdNpp5/5LBXlU2PVmqpVr7mWxk0ajn9LzULzllvp1FFKPXbtqVv3nnsZqLUhIw4dadjwkUe5U6N9bJ+o0Ttyn1OjTW0Si0vPHtQgNruZoHmd6GQGYhwJxG0SQEHzZHY4xci" +
    "T3GR2ZMahUAY10gmn0SQGgrET66A7uwe5T7kFjb/FjX9FLkx0f4NcmOg2uY/cXlBrZf1HkQVonsKZ00MGLrYmhR1/uI6/MIavGvhv6B8bElRJDj8BKiwW0vp4gF0AAAAGYk" +
    "tHRAD/AP8A/6C9p5MAAAAJcEhZcwAADsIAAA7CARUoSoAAAAAHdElNRQfkChMLECg6P0auAAAA20lEQVQ4y41U2w6FIAxri57/T/hgfUHPwF0gIWoyy3oZ7L1jLOK/OLYAn" +
    "AAOwV80T0WFn6IKkQDa+G5ZocZzQmRydPMQbVEz+4yOljmaAKSAxPN+jh6bEsaqBOfCWgB+lTMPqpRoaPfUoyVyGYkYIa4/HZWFqTO2aPJ7N4+s8vjuqhCePBGaKnlkdCxH" +
    "gZk8npUpGTsmyhBhoxaFQkuPqixkpKM3Ci9rb6bXHqnkJmOU8Ah10qqamyOSB5ZIxXoitBMz15loZribnjIUYXoUeP0pvEzBtZNHrjfuDRr4BoNUBycoAAAAAElFTkSuQmCC";

const rightCornerBottomImage =
    "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAoAAAAqCAYAAACKnsqLAAACznpUWHRSYXcgcHJvZmlsZSB0eXBlIGV4aWYAAHja7ZVtctwgDIb/c4oewZIQE" +
    "sfBfMz0Bj1+XzC7m022mWnTmf6piQFrhST0CBL6j+8jfMNDlI4Q1TzllA48McfMBRM/ruca6Yirv56yR3qWB779wBAJRrk+U9/6BXJ9LLC45eezPFjddnwbunneBmV6Zky2" +
    "nm9Dwpec9nfIt4jjm+3sN9Fhe9EV5bvvaEhGU9gTDtyF5ECfpoJcb5my1WcoEWRzntCr+OvcBamvk3efvcvdUbZcnlMRjrQV0rscbTnp69ytDL2NiB6en34wvi/5kLsxmo/" +
    "Rr92VmJCpFPambltZMyieSKWsZQnN8CrmtlpGc2yxglgDzROtBsrEyOygSI0KDeprrFQRYuTOCI2ZK8uSuRhnrjIRxNlosEmWFsTBpIKaQMz3WGj5zctfJYfnRtBkgrFJ8U" +
    "MLr4R/0u6Gxph1QHT4lSeUBeLiWXoIY5KbPbQAhMbOqa78rhbe1M3xBqyAoK40OzZYjvMycSo9aksWZ4GeHjEc19Ega9sAUgTfimBIQADnQpTm4WA2IuTRwacgcpbIJwiQK" +
    "jcKA2xEEuA4T99YY7R0WfkS42oBCMURMaDJUgArRkX9WHTUUFHRGFQ1qalr1pIkxaQpJUvzjiomFk0tmZlbtuLi0dWTm7tnL5mz4ArTnLKF7DnnUuC0wHTB6gKNUk4+5Yyn" +
    "num00898loryqbFqTdWq11xL4yYNx7+lZqF5y6106iilHrv21K17z70M1NqQEYeONGz4yKPcqdE+tk/U6B25z6nRpjaJxaVnD2oQm91M0LxOdDIDMY4E4jYJoKB5MjucYuR" +
    "JbjI7MuNQKIMa6YTTaBIDwdiJddCd3YPcp9yCxt/ixr8iFya6v0EuTHSb3EduL6i1sv6jyAI0T+HM6SEDF1uTwo4/XMdfGMNXDfw39I8NCaokh58yDBbTmnsE1AAAAAZiS0" +
    "dEAP8A/wD/oL2nkwAAAAlwSFlzAAAOwgAADsIBFShKgAAAAAd0SU1FB+QKEwsRABaR3xUAAADzSURBVDjLlZSxDsMgDEQfNB0qdejSoVL/On/bYrtLUlkEMEFCJOI4+3yGt" +
    "K5rAhJwAa7ADbgDD+AJvIB3pj0M0G01QGtgcpvJH6qBnsUcs/ZCH9LIEaDFaG61WlBusNDKM3dCWi0wBwDZZ54RApRe6MPIjc1DsffQNmDrlocWW09My2vplcc3h8w0hUbO" +
    "7KDT3WMjZ3yOoYWyfZeZNtNIjBclEdBGFtZgjbz2d6ZE5dGZ8uzhy6gpkiv20GutpkSqNSq4V6zAZ+YqfIGyDK6qbaBllKNGL663Ttz89h4Azjjz93nGa9n+ZemAkmMtQJ7" +
    "tx/IDWBnMMlnzWL0AAAAASUVORK5CYII=";

// styles
export const fullHeightStyle = b.styleDef(
    {
        height: "100%",
    },
    {},
    "scroll-full-height"
);

export const wrapperStyle = b.styleDef(
    {
        position: "relative",
    },
    {},
    "scroll-wrapper"
);

export const wrapperZIndexResetStyle = b.styleDefEx(
    wrapperStyle,
    {
        zIndex: 0,
    },
    {},
    "scroll-wrapper-z-index-reset"
);

export const overflowHiddenStyle = b.styleDef({
    overflow: "hidden !important",
});

export const contentStyle = b.styleDef(
    {
        overflow: "auto",
    },
    {},
    "scroll-content"
);

export const contentRelativeStyle = b.styleDef(
    {
        position: "relative",
    },
    {},
    "scroll-content-relative"
);

export const contentSidePaddingStyle = b.styleDef(
    {
        paddingLeft: sidePadding,
        paddingRight: sidePadding,
    },
    {},
    "scroll-content-with-padding"
);

export const contentSidePaddingForAsterisksStyle = b.styleDef(
    {
        marginLeft: -1 * sidePaddingForAsterisks,
        paddingLeft: sidePaddingForAsterisks,
    },
    {},
    "scroll-content-with-padding-for-asterisks"
);

export const shadowStyle = b.styleDef(
    {
        position: "absolute",
        opacity: 0,
        height: shadowHeight,
        marginLeft: shadowBoundary,
        marginRight: shadowBoundary,
        backgroundRepeat: "repeat-x",
        pointerEvents: "none",
        zIndex: 10,
    },
    {
        before: [
            {
                position: "absolute",
                display: "block",
                content: "",
                top: 0,
                left: -shadowBoundary,
                width: shadowBoundary,
                height: shadowHeight,
                pointerEvents: "none",
                zIndex: 10,
            },
        ],
        after: [
            {
                position: "absolute",
                display: "block",
                content: "",
                top: 0,
                right: -shadowBoundary,
                width: shadowBoundary,
                height: shadowHeight,
                pointerEvents: "none",
                zIndex: 10,
            },
        ],
    },
    "scroll-shadow"
);

export const topShadowStyle = b.styleDef(
    {
        top: 0,
        backgroundImage: "url(" + topShadowImage + ")",
    },
    {
        before: {
            backgroundImage: "url(" + topCornerLeftImage + ")",
        },
        after: {
            backgroundImage: "url(" + topCornerRightImage + ")",
        },
    },
    "scroll-shadow-top"
);

export const bottomShadowStyle = b.styleDef(
    {
        bottom: 0,
        backgroundImage: "url(" + bottomShadowImage + ")",
    },
    {
        before: {
            backgroundImage: "url(" + bottomCornerLeftImage + ")",
        },
        after: {
            backgroundImage: "url(" + bottomCornerRightImage + ")",
        },
    },
    "scroll-shadow-bottom"
);

export const shadowVerticalStyle = b.styleDefEx(
    shadowStyle,
    {
        width: shadowHeight,
        height: undefined,
        marginLeft: 0,
        marginRight: 0,
        marginTop: shadowBoundary,
        marginBottom: shadowBoundary,
        backgroundRepeat: "repeat-y",
    },
    {
        before: [
            {
                left: 0,
                top: -shadowBoundary,
                height: shadowBoundary,
                width: shadowHeight,
            },
        ],
        after: [
            {
                left: 0,
                top: "auto",
                bottom: -shadowBoundary,
                height: shadowBoundary,
                width: shadowHeight,
            },
        ],
    },
    "scroll-shadow-vertical"
);

export const leftShadowStyle = b.styleDef(
    {
        left: 0,
        top: 0,
        backgroundImage: "url(" + leftShadowImage + ")",
    },
    {
        before: {
            backgroundImage: "url(" + leftCornerTopImage + ")",
        },
        after: {
            backgroundImage: "url(" + leftCornerBottomImage + ")",
        },
    },
    "scroll-shadow-left"
);

export const rightShadowStyle = b.styleDef(
    {
        right: 0,
        top: 0,
        backgroundImage: "url(" + rightShadowImage + ")",
    },
    {
        before: {
            backgroundImage: "url(" + rightCornerTopImage + ")",
        },
        after: {
            backgroundImage: "url(" + rightCornerBottomImage + ")",
        },
    },
    "scroll-shadow-right"
);

// just for identification in DOM
export const scrollShadowWithOffsetWrapperStyle = b.styleDef({});

export const scrollShadowWithOffsetWrapperFullSizeStyle = b.styleDef({
    height: "100%",
});

export const scrollShadowOffsetIECorrection = b.styleDef({ paddingBottom: 1, marginBottom: -1 });

export const shadowLeftSidePaddingStyle = b.styleDefEx(shadowStyle, {
    left: sidePadding,
});

export const scrollShadowBottomBorderStyle = b.styleDef(
    {
        position: "absolute",
        bottom: -bottomBorderHeight,
        left: 0,
        right: 0,
        height: 1,
        background: Color.Grey300,
    },
    {},
    "scroll-shadow-bottom-border"
);

export const scrollShadowWithBottomBorderStyle = b.styleDef(
    {
        paddingBottom: bottomBorderHeight,
        boxSizing: "border-box",
    },
    {},
    "scroll-shadow-with-bottom-border"
);
